{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import iglob\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import screed\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Change to Quest for Orthologs 2019 data directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019\n"
     ]
    }
   ],
   "source": [
    "cd ~/data_sm/kmer-hashing/quest-for-orthologs/data/2019/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 2.7G\n",
      "drwxr-xr-x  5 olga root 4.0K Jan 17 07:54 \u001b[0m\u001b[01;34m.\u001b[0m/\n",
      "drwxr-xr-x  3 olga root 4.0K Dec 25 17:48 \u001b[01;34m..\u001b[0m/\n",
      "drwxr-xr-x  5 olga czb  4.0K Dec 26 19:44 \u001b[01;34mArchaea\u001b[0m/\n",
      "drwxr-xr-x  5 olga czb   16K Dec 26 19:44 \u001b[01;34mBacteria\u001b[0m/\n",
      "drwxr-xr-x 11 olga czb   32K Jan 17 07:52 \u001b[01;34mEukaryota\u001b[0m/\n",
      "-rw-r--r--  1 olga czb  754K Jan 10 07:50 human_transcription_factors_with_uniprot_ids.csv\n",
      "-rw-r--r--  1 olga czb   68K Jan 10 07:50 \u001b[01;31mhuman_transcription_factors_with_uniprot_ids.csv.gz\u001b[0m\n",
      "-rw-r--r--  1 olga czb  133K Jan 10 07:50 human_transcription_factors_with_uniprot_ids.parquet\n",
      "-rw-r--r--  1 olga czb   14K Jan 17 07:54 human_transcription_factors_with_uniprot_ids_random_subset100.csv\n",
      "-rw-r--r--  1 olga czb  3.5K Jan 17 07:54 \u001b[01;31mhuman_transcription_factors_with_uniprot_ids_random_subset100.csv.gz\u001b[0m\u001b[K\n",
      "-rw-r--r--  1 olga czb   12K Jan 17 07:54 human_transcription_factors_with_uniprot_ids_random_subset100.parquet\n",
      "-rw-r--r--  1 olga czb   76K Jan 15 10:57 human_visual_transduction_with_uniprot_ids.csv\n",
      "-rw-r--r--  1 olga czb   19K Jan 15 10:57 \u001b[01;31mhuman_visual_transduction_with_uniprot_ids.csv.gz\u001b[0m\n",
      "-rw-r--r--  1 olga czb   28K Jan 15 10:57 human_visual_transduction_with_uniprot_ids.parquet\n",
      "-rw-r--r--  1 olga czb   64M Jan 10 07:43 opisthokont_not_human_transcription_factors_ensembl_compara.csv\n",
      "-rw-r--r--  1 olga czb   19M Jan 15 18:51 \u001b[01;31mopisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.csv.gz\u001b[0m\u001b[K\n",
      "-rw-r--r--  1 olga czb   23M Jan 15 18:51 opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.parquet\n",
      "-rw-r--r--  1 olga czb   12M Jan 10 07:43 opisthokont_not_human_transcription_factors_ensembl_compara.parquet\n",
      "-rw-r--r--  1 olga czb  3.4M Jan 15 18:10 opisthokont_not_human_visual_system_ensembl_compara.csv\n",
      "-rw-r--r--  1 olga czb  431K Jan 15 18:10 opisthokont_not_human_visual_system_ensembl_compara.parquet\n",
      "-rw-r--r--  1 olga czb   17M Jan 15 18:49 \u001b[01;31mopisthokont_not_human_visual_transduction_ensembl_compara_merged_uniprot.csv.gz\u001b[0m\u001b[K\n",
      "-rw-r--r--  1 olga czb   21M Jan 15 18:50 opisthokont_not_human_visual_transduction_ensembl_compara_merged_uniprot.parquet\n",
      "-rw-r--r--  1 olga czb   661 Jan  8 17:33 qfo_human_vs_opisthokont_tfs.sh\n",
      "-rw-r--r--  1 olga czb  2.6G Dec 25 18:46 \u001b[01;31mQfO_release_2019_04.tar.gz\u001b[0m\n",
      "-rw-r--r--  1 olga czb   18K May 10  2019 README\n",
      "-rw-r--r--  1 olga czb   12K Jan  8 07:47 species_metadata.csv\n"
     ]
    }
   ],
   "source": [
    "ls -lha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34mhuman-protein-fastas\u001b[0m/\n",
      "\u001b[01;34mhuman-protein-fastas-concatenated\u001b[0m/\n",
      "\u001b[01;34mhuman-transcription-factor-fastas\u001b[0m/\n",
      "\u001b[01;34mhuman-transcription-factor-fastas-random-subset100\u001b[0m/\n",
      "\u001b[01;34mhuman-visual-transduction-fastas\u001b[0m/\n",
      "\u001b[01;34mnot-human-protein-fastas\u001b[0m/\n",
      "\u001b[01;34mnot-human-transcription-factor-fastas\u001b[0m/\n",
      "\u001b[01;34mnot-human-visual-transduction-fastas\u001b[0m/\n",
      "\u001b[01;34mprotein-fastas\u001b[0m/\n",
      "UP000000437_7955_additional.fasta\n",
      "UP000000437_7955_DNA.fasta\n",
      "UP000000437_7955_DNA.miss\n",
      "UP000000437_7955.fasta\n",
      "UP000000437_7955_gcoord.xml\n",
      "UP000000437_7955.gene2acc\n",
      "UP000000437_7955.idmapping\n",
      "UP000000437_7955.xml\n",
      "UP000000539_9031_additional.fasta\n",
      "UP000000539_9031_DNA.fasta\n",
      "UP000000539_9031_DNA.miss\n",
      "UP000000539_9031.fasta\n",
      "UP000000539_9031_gcoord.xml\n",
      "UP000000539_9031.gene2acc\n",
      "UP000000539_9031.idmapping\n",
      "UP000000539_9031.xml\n",
      "UP000000542_5664_DNA.fasta\n",
      "UP000000542_5664.fasta\n",
      "UP000000542_5664.gene2acc\n",
      "UP000000542_5664.idmapping\n",
      "UP000000542_5664.xml\n",
      "UP000000559_237561_additional.fasta\n",
      "UP000000559_237561_DNA.fasta\n",
      "UP000000559_237561_DNA.miss\n",
      "UP000000559_237561.fasta\n",
      "UP000000559_237561.gene2acc\n",
      "UP000000559_237561.idmapping\n",
      "UP000000559_237561.xml\n",
      "UP000000561_237631_additional.fasta\n",
      "UP000000561_237631_DNA.fasta\n",
      "UP000000561_237631.fasta\n",
      "UP000000561_237631.gene2acc\n",
      "UP000000561_237631.idmapping\n",
      "UP000000561_237631.xml\n",
      "UP000000589_10090_additional.fasta\n",
      "UP000000589_10090_DNA.fasta\n",
      "UP000000589_10090_DNA.miss\n",
      "UP000000589_10090.fasta\n",
      "UP000000589_10090_gcoord.xml\n",
      "UP000000589_10090.gene2acc\n",
      "UP000000589_10090.idmapping\n",
      "UP000000589_10090.xml\n",
      "UP000000600_5888_DNA.fasta\n",
      "UP000000600_5888.fasta\n",
      "UP000000600_5888.gene2acc\n",
      "UP000000600_5888.idmapping\n",
      "UP000000600_5888.xml\n",
      "UP000000803_7227_additional.fasta\n",
      "UP000000803_7227_DNA.fasta\n",
      "UP000000803_7227.fasta\n",
      "UP000000803_7227.gene2acc\n",
      "UP000000803_7227.idmapping\n",
      "UP000000803_7227.xml\n",
      "UP000001038_8090_additional.fasta\n",
      "UP000001038_8090_DNA.fasta\n",
      "UP000001038_8090_DNA.miss\n",
      "UP000001038_8090.fasta\n",
      "UP000001038_8090_gcoord.xml\n",
      "UP000001038_8090.gene2acc\n",
      "UP000001038_8090.idmapping\n",
      "UP000001038_8090.xml\n",
      "UP000001055_321614_DNA.fasta\n",
      "UP000001055_321614.fasta\n",
      "UP000001055_321614.gene2acc\n",
      "UP000001055_321614.idmapping\n",
      "UP000001055_321614.xml\n",
      "UP000001300_284591_additional.fasta\n",
      "UP000001300_284591_DNA.fasta\n",
      "UP000001300_284591_DNA.miss\n",
      "UP000001300_284591.fasta\n",
      "UP000001300_284591.gene2acc\n",
      "UP000001300_284591.idmapping\n",
      "UP000001300_284591.xml\n",
      "UP000001312_665079_DNA.fasta\n",
      "UP000001312_665079.fasta\n",
      "UP000001312_665079.gene2acc\n",
      "UP000001312_665079.idmapping\n",
      "UP000001312_665079.xml\n",
      "UP000001357_81824_DNA.fasta\n",
      "UP000001357_81824.fasta\n",
      "UP000001357_81824.gene2acc\n",
      "UP000001357_81824.idmapping\n",
      "UP000001357_81824.xml\n",
      "UP000001449_35128_additional.fasta\n",
      "UP000001449_35128_DNA.fasta\n",
      "UP000001449_35128.fasta\n",
      "UP000001449_35128.gene2acc\n",
      "UP000001449_35128.idmapping\n",
      "UP000001449_35128.xml\n",
      "UP000001450_36329_additional.fasta\n",
      "UP000001450_36329_DNA.fasta\n",
      "UP000001450_36329_DNA.miss\n",
      "UP000001450_36329.fasta\n",
      "UP000001450_36329.gene2acc\n",
      "UP000001450_36329.idmapping\n",
      "UP000001450_36329.xml\n",
      "UP000001519_9595_additional.fasta\n",
      "UP000001519_9595_DNA.fasta\n",
      "UP000001519_9595_DNA.miss\n",
      "UP000001519_9595.fasta\n",
      "UP000001519_9595_gcoord.xml\n",
      "UP000001519_9595.gene2acc\n",
      "UP000001519_9595.idmapping\n",
      "UP000001519_9595.xml\n",
      "UP000001542_5722_DNA.fasta\n",
      "UP000001542_5722.fasta\n",
      "UP000001542_5722.gene2acc\n",
      "UP000001542_5722.idmapping\n",
      "UP000001542_5722.xml\n",
      "UP000001548_184922_DNA.fasta\n",
      "UP000001548_184922.fasta\n",
      "UP000001548_184922.gene2acc\n",
      "UP000001548_184922.idmapping\n",
      "UP000001548_184922.xml\n",
      "UP000001554_7739_additional.fasta\n",
      "UP000001554_7739_DNA.fasta\n",
      "UP000001554_7739_DNA.miss\n",
      "UP000001554_7739.fasta\n",
      "UP000001554_7739.gene2acc\n",
      "UP000001554_7739.idmapping\n",
      "UP000001554_7739.xml\n",
      "UP000001555_6945_DNA.fasta\n",
      "UP000001555_6945.fasta\n",
      "UP000001555_6945_gcoord.xml\n",
      "UP000001555_6945.gene2acc\n",
      "UP000001555_6945.idmapping\n",
      "UP000001555_6945.xml\n",
      "UP000001593_45351_additional.fasta\n",
      "UP000001593_45351_DNA.fasta\n",
      "UP000001593_45351.fasta\n",
      "UP000001593_45351.gene2acc\n",
      "UP000001593_45351.idmapping\n",
      "UP000001593_45351.xml\n",
      "UP000001805_367110_additional.fasta\n",
      "UP000001805_367110_DNA.fasta\n",
      "UP000001805_367110.fasta\n",
      "UP000001805_367110.gene2acc\n",
      "UP000001805_367110.idmapping\n",
      "UP000001805_367110.xml\n",
      "UP000001940_6239_additional.fasta\n",
      "UP000001940_6239_DNA.fasta\n",
      "UP000001940_6239_DNA.miss\n",
      "UP000001940_6239.fasta\n",
      "UP000001940_6239.gene2acc\n",
      "UP000001940_6239.idmapping\n",
      "UP000001940_6239.xml\n",
      "UP000002149_214684_additional.fasta\n",
      "UP000002149_214684_DNA.fasta\n",
      "UP000002149_214684_DNA.miss\n",
      "UP000002149_214684.fasta\n",
      "UP000002149_214684.gene2acc\n",
      "UP000002149_214684.idmapping\n",
      "UP000002149_214684.xml\n",
      "UP000002195_44689_additional.fasta\n",
      "UP000002195_44689_DNA.fasta\n",
      "UP000002195_44689_DNA.miss\n",
      "UP000002195_44689.fasta\n",
      "UP000002195_44689.gene2acc\n",
      "UP000002195_44689.idmapping\n",
      "UP000002195_44689.xml\n",
      "UP000002254_9615_additional.fasta\n",
      "UP000002254_9615_DNA.fasta\n",
      "UP000002254_9615_DNA.miss\n",
      "UP000002254_9615.fasta\n",
      "UP000002254_9615_gcoord.xml\n",
      "UP000002254_9615.gene2acc\n",
      "UP000002254_9615.idmapping\n",
      "UP000002254_9615.xml\n",
      "UP000002277_9598_additional.fasta\n",
      "UP000002277_9598_DNA.fasta\n",
      "UP000002277_9598_DNA.miss\n",
      "UP000002277_9598.fasta\n",
      "UP000002277_9598_gcoord.xml\n",
      "UP000002277_9598.gene2acc\n",
      "UP000002277_9598.idmapping\n",
      "UP000002277_9598.xml\n",
      "UP000002280_13616_additional.fasta\n",
      "UP000002280_13616_DNA.fasta\n",
      "UP000002280_13616_DNA.miss\n",
      "UP000002280_13616.fasta\n",
      "UP000002280_13616_gcoord.xml\n",
      "UP000002280_13616.gene2acc\n",
      "UP000002280_13616.idmapping\n",
      "UP000002280_13616.xml\n",
      "UP000002311_559292_additional.fasta\n",
      "UP000002311_559292_DNA.fasta\n",
      "UP000002311_559292_DNA.miss\n",
      "UP000002311_559292.fasta\n",
      "UP000002311_559292_gcoord.xml\n",
      "UP000002311_559292.gene2acc\n",
      "UP000002311_559292.idmapping\n",
      "UP000002311_559292.xml\n",
      "UP000002485_284812_additional.fasta\n",
      "UP000002485_284812_DNA.fasta\n",
      "UP000002485_284812_DNA.miss\n",
      "UP000002485_284812.fasta\n",
      "UP000002485_284812.gene2acc\n",
      "UP000002485_284812.idmapping\n",
      "UP000002485_284812.xml\n",
      "UP000002494_10116_additional.fasta\n",
      "UP000002494_10116_DNA.fasta\n",
      "UP000002494_10116_DNA.miss\n",
      "UP000002494_10116.fasta\n",
      "UP000002494_10116_gcoord.xml\n",
      "UP000002494_10116.gene2acc\n",
      "UP000002494_10116.idmapping\n",
      "UP000002494_10116.xml\n",
      "UP000002530_330879_DNA.fasta\n",
      "UP000002530_330879.fasta\n",
      "UP000002530_330879.gene2acc\n",
      "UP000002530_330879.idmapping\n",
      "UP000002530_330879.xml\n",
      "UP000005238_164328_DNA.fasta\n",
      "UP000005238_164328.fasta\n",
      "UP000005238_164328_gcoord.xml\n",
      "UP000005238_164328.gene2acc\n",
      "UP000005238_164328.idmapping\n",
      "UP000005238_164328.xml\n",
      "UP000005640_9606_additional.fasta\n",
      "UP000005640_9606_DNA.fasta\n",
      "UP000005640_9606_DNA.miss\n",
      "UP000005640_9606.fasta\n",
      "UP000005640_9606_gcoord.miss\n",
      "UP000005640_9606_gcoord.xml\n",
      "UP000005640_9606.gene2acc\n",
      "UP000005640_9606.idmapping\n",
      "UP000005640_9606.xml\n",
      "UP000006548_3702_additional.fasta\n",
      "UP000006548_3702_DNA.fasta\n",
      "UP000006548_3702_DNA.miss\n",
      "UP000006548_3702.fasta\n",
      "UP000006548_3702_gcoord.xml\n",
      "UP000006548_3702.gene2acc\n",
      "UP000006548_3702.idmapping\n",
      "UP000006548_3702.xml\n",
      "UP000006727_3218_additional.fasta\n",
      "UP000006727_3218_DNA.fasta\n",
      "UP000006727_3218.fasta\n",
      "UP000006727_3218.gene2acc\n",
      "UP000006727_3218.idmapping\n",
      "UP000006727_3218.xml\n",
      "UP000006906_3055_additional.fasta\n",
      "UP000006906_3055_DNA.fasta\n",
      "UP000006906_3055_DNA.miss\n",
      "UP000006906_3055.fasta\n",
      "UP000006906_3055.gene2acc\n",
      "UP000006906_3055.idmapping\n",
      "UP000006906_3055.xml\n",
      "UP000007062_7165_additional.fasta\n",
      "UP000007062_7165_DNA.fasta\n",
      "UP000007062_7165_DNA.miss\n",
      "UP000007062_7165.fasta\n",
      "UP000007062_7165_gcoord.xml\n",
      "UP000007062_7165.gene2acc\n",
      "UP000007062_7165.idmapping\n",
      "UP000007062_7165.xml\n",
      "UP000007241_684364_DNA.fasta\n",
      "UP000007241_684364.fasta\n",
      "UP000007241_684364.gene2acc\n",
      "UP000007241_684364.idmapping\n",
      "UP000007241_684364.xml\n",
      "UP000007266_7070_additional.fasta\n",
      "UP000007266_7070_DNA.fasta\n",
      "UP000007266_7070.fasta\n",
      "UP000007266_7070.gene2acc\n",
      "UP000007266_7070.idmapping\n",
      "UP000007266_7070.xml\n",
      "UP000007305_4577_additional.fasta\n",
      "UP000007305_4577_DNA.fasta\n",
      "UP000007305_4577_DNA.miss\n",
      "UP000007305_4577.fasta\n",
      "UP000007305_4577_gcoord.xml\n",
      "UP000007305_4577.gene2acc\n",
      "UP000007305_4577.idmapping\n",
      "UP000007305_4577.xml\n",
      "UP000008143_8364_additional.fasta\n",
      "UP000008143_8364_DNA.fasta\n",
      "UP000008143_8364_DNA.miss\n",
      "UP000008143_8364.fasta\n",
      "UP000008143_8364_gcoord.xml\n",
      "UP000008143_8364.gene2acc\n",
      "UP000008143_8364.idmapping\n",
      "UP000008143_8364.xml\n",
      "UP000008144_7719_additional.fasta\n",
      "UP000008144_7719_DNA.fasta\n",
      "UP000008144_7719.fasta\n",
      "UP000008144_7719_gcoord.xml\n",
      "UP000008144_7719.gene2acc\n",
      "UP000008144_7719.idmapping\n",
      "UP000008144_7719.xml\n",
      "UP000008783_418459_additional.fasta\n",
      "UP000008783_418459_DNA.fasta\n",
      "UP000008783_418459.fasta\n",
      "UP000008783_418459.gene2acc\n",
      "UP000008783_418459.idmapping\n",
      "UP000008783_418459.xml\n",
      "UP000009136_9913_additional.fasta\n",
      "UP000009136_9913_DNA.fasta\n",
      "UP000009136_9913_DNA.miss\n",
      "UP000009136_9913.fasta\n",
      "UP000009136_9913_gcoord.xml\n",
      "UP000009136_9913.gene2acc\n",
      "UP000009136_9913.idmapping\n",
      "UP000009136_9913.xml\n",
      "UP000015101_6412_DNA.fasta\n",
      "UP000015101_6412.fasta\n",
      "UP000015101_6412_gcoord.xml\n",
      "UP000015101_6412.gene2acc\n",
      "UP000015101_6412.idmapping\n",
      "UP000015101_6412.xml\n",
      "UP000018468_7918_additional.fasta\n",
      "UP000018468_7918_DNA.fasta\n",
      "UP000018468_7918.fasta\n",
      "UP000018468_7918_gcoord.xml\n",
      "UP000018468_7918.gene2acc\n",
      "UP000018468_7918.idmapping\n",
      "UP000018468_7918.xml\n",
      "UP000059680_39947_additional.fasta\n",
      "UP000059680_39947_DNA.fasta\n",
      "UP000059680_39947_DNA.miss\n",
      "UP000059680_39947.fasta\n",
      "UP000059680_39947.gene2acc\n",
      "UP000059680_39947.idmapping\n",
      "UP000059680_39947.xml\n"
     ]
    }
   ],
   "source": [
    "ls Eukaryota/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download orthology and transcription factor data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read orthologous transcription factors "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(343153, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>target__perc_pos</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>24.8792</td>\n",
       "      <td>37.1981</td>\n",
       "      <td>ENSP00000467286</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000175787</td>\n",
       "      <td>17.0813</td>\n",
       "      <td>25.5390</td>\n",
       "      <td>ENSP00000378792</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>25.6039</td>\n",
       "      <td>37.1981</td>\n",
       "      <td>ENSP00000467286</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000187815</td>\n",
       "      <td>20.1521</td>\n",
       "      <td>29.2776</td>\n",
       "      <td>ENSP00000361791</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.35552</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>26.3285</td>\n",
       "      <td>38.1642</td>\n",
       "      <td>ENSP00000467286</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000172006</td>\n",
       "      <td>20.2602</td>\n",
       "      <td>29.3680</td>\n",
       "      <td>ENSP00000321132</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.33996</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>27.2947</td>\n",
       "      <td>38.8889</td>\n",
       "      <td>ENSP00000467286</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000188629</td>\n",
       "      <td>23.4927</td>\n",
       "      <td>33.4719</td>\n",
       "      <td>ENSP00000415070</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.18553</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>20.2899</td>\n",
       "      <td>25.8454</td>\n",
       "      <td>ENSP00000467286</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000270011</td>\n",
       "      <td>26.1682</td>\n",
       "      <td>33.3333</td>\n",
       "      <td>ENSP00000445323</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     dn_ds    method_link_type       source__id  source__perc_id  \\\n",
       "0      NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   \n",
       "1      NaN  ENSEMBL_PARALOGUES  ENSG00000267179          25.6039   \n",
       "2  0.35552  ENSEMBL_PARALOGUES  ENSG00000267179          26.3285   \n",
       "3  0.33996  ENSEMBL_PARALOGUES  ENSG00000267179          27.2947   \n",
       "4  0.18553  ENSEMBL_PARALOGUES  ENSG00000267179          20.2899   \n",
       "\n",
       "   source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "0           37.1981    ENSP00000467286    homo_sapiens              9606   \n",
       "1           37.1981    ENSP00000467286    homo_sapiens              9606   \n",
       "2           38.1642    ENSP00000467286    homo_sapiens              9606   \n",
       "3           38.8889    ENSP00000467286    homo_sapiens              9606   \n",
       "4           25.8454    ENSP00000467286    homo_sapiens              9606   \n",
       "\n",
       "        target__id  target__perc_id  target__perc_pos target__protein_id  \\\n",
       "0  ENSG00000175787          17.0813           25.5390    ENSP00000378792   \n",
       "1  ENSG00000187815          20.1521           29.2776    ENSP00000361791   \n",
       "2  ENSG00000172006          20.2602           29.3680    ENSP00000321132   \n",
       "3  ENSG00000188629          23.4927           33.4719    ENSP00000415070   \n",
       "4  ENSG00000270011          26.1682           33.3333    ENSP00000445323   \n",
       "\n",
       "  target__species  target__taxon_id taxonomy_level           type  \n",
       "0    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "1    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "2    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "3    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "4    homo_sapiens              9606      Bilateria  other_paralog  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs_original = pd.read_csv('opisthokont_not_human_transcription_factors_ensembl_compara.csv')\n",
    "print(tfs_original.shape)\n",
    "tfs_original.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read random subset of TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(100, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>ID</th>\n",
       "      <th>DBD</th>\n",
       "      <th>Motif status (Feb 2018)(Link to human TFs annotation)</th>\n",
       "      <th>IUPAC consensus (from selected PWM)</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>id_type</th>\n",
       "      <th>db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TFCP2L1</td>\n",
       "      <td>ENSG00000115112</td>\n",
       "      <td>Grainyhead</td>\n",
       "      <td>Known motif – High-throughput in vitro [942]</td>\n",
       "      <td>DCYRGHNNNDDCYRGH</td>\n",
       "      <td>Q9NZI6</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000115112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PREB</td>\n",
       "      <td>ENSG00000138073</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Likely sequence specific TF according to liter...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H7C0W3</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000138073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>ENSG00000171634</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Known motif – In vivo/Misc source [79]</td>\n",
       "      <td>KKKNTTGTKKNV</td>\n",
       "      <td>J3KSY0</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000171634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MXI1</td>\n",
       "      <td>ENSG00000119950</td>\n",
       "      <td>bHLH</td>\n",
       "      <td>Known motif – In vivo/Misc source [553]</td>\n",
       "      <td>CCACGTGG</td>\n",
       "      <td>P50539</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000119950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HNF4A</td>\n",
       "      <td>ENSG00000101076</td>\n",
       "      <td>Nuclear receptor</td>\n",
       "      <td>Known motif – High-throughput in vitro [369]</td>\n",
       "      <td>VRGGTCAAAGTCCA</td>\n",
       "      <td>F8WBS7</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000101076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene               ID               DBD  \\\n",
       "0  TFCP2L1  ENSG00000115112        Grainyhead   \n",
       "1     PREB  ENSG00000138073           Unknown   \n",
       "2     BPTF  ENSG00000171634           Unknown   \n",
       "3     MXI1  ENSG00000119950              bHLH   \n",
       "4    HNF4A  ENSG00000101076  Nuclear receptor   \n",
       "\n",
       "  Motif status (Feb 2018)(Link to human TFs annotation)  \\\n",
       "0       Known motif – High-throughput in vitro [942]      \n",
       "1  Likely sequence specific TF according to liter...      \n",
       "2             Known motif – In vivo/Misc source [79]      \n",
       "3            Known motif – In vivo/Misc source [553]      \n",
       "4       Known motif – High-throughput in vitro [369]      \n",
       "\n",
       "  IUPAC consensus (from selected PWM) uniprot_id  id_type            db_id  \n",
       "0                    DCYRGHNNNDDCYRGH     Q9NZI6  Ensembl  ENSG00000115112  \n",
       "1                                 NaN     H7C0W3  Ensembl  ENSG00000138073  \n",
       "2                        KKKNTTGTKKNV     J3KSY0  Ensembl  ENSG00000171634  \n",
       "3                            CCACGTGG     P50539  Ensembl  ENSG00000119950  \n",
       "4                      VRGGTCAAAGTCCA     F8WBS7  Ensembl  ENSG00000101076  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs_random_subset = pd.read_csv('human_transcription_factors_with_uniprot_ids_random_subset100.csv')\n",
    "print(tfs_random_subset.shape)\n",
    "tfs_random_subset.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset TF orthology to the random subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20586, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>target__perc_pos</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1406</th>\n",
       "      <td>0.15000</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>96.0212</td>\n",
       "      <td>96.1096</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>99.6330</td>\n",
       "      <td>99.7248</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>gorilla_gorilla</td>\n",
       "      <td>9595</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>0.19118</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>99.2042</td>\n",
       "      <td>99.2927</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPPAG00000028735</td>\n",
       "      <td>99.7333</td>\n",
       "      <td>99.8222</td>\n",
       "      <td>ENSPPAP00000011050</td>\n",
       "      <td>pan_paniscus</td>\n",
       "      <td>9597</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1408</th>\n",
       "      <td>0.08163</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>99.2927</td>\n",
       "      <td>99.4695</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPTRG00000010138</td>\n",
       "      <td>99.8222</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>ENSPTRP00000017225</td>\n",
       "      <td>pan_troglodytes</td>\n",
       "      <td>9598</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1409</th>\n",
       "      <td>0.15306</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>98.2317</td>\n",
       "      <td>98.6737</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPPYG00000009265</td>\n",
       "      <td>98.7556</td>\n",
       "      <td>99.2000</td>\n",
       "      <td>ENSPPYP00000010399</td>\n",
       "      <td>pongo_abelii</td>\n",
       "      <td>9601</td>\n",
       "      <td>Hominidae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1410</th>\n",
       "      <td>0.24272</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>97.4359</td>\n",
       "      <td>98.3201</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSNLEG00000008406</td>\n",
       "      <td>97.9556</td>\n",
       "      <td>98.8444</td>\n",
       "      <td>ENSNLEP00000010232</td>\n",
       "      <td>nomascus_leucogenys</td>\n",
       "      <td>61853</td>\n",
       "      <td>Hominoidea</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dn_ds     method_link_type       source__id  source__perc_id  \\\n",
       "1406  0.15000  ENSEMBL_ORTHOLOGUES  ENSG00000101544          96.0212   \n",
       "1407  0.19118  ENSEMBL_ORTHOLOGUES  ENSG00000101544          99.2042   \n",
       "1408  0.08163  ENSEMBL_ORTHOLOGUES  ENSG00000101544          99.2927   \n",
       "1409  0.15306  ENSEMBL_ORTHOLOGUES  ENSG00000101544          98.2317   \n",
       "1410  0.24272  ENSEMBL_ORTHOLOGUES  ENSG00000101544          97.4359   \n",
       "\n",
       "      source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "1406           96.1096    ENSP00000262198    homo_sapiens              9606   \n",
       "1407           99.2927    ENSP00000262198    homo_sapiens              9606   \n",
       "1408           99.4695    ENSP00000262198    homo_sapiens              9606   \n",
       "1409           98.6737    ENSP00000262198    homo_sapiens              9606   \n",
       "1410           98.3201    ENSP00000262198    homo_sapiens              9606   \n",
       "\n",
       "              target__id  target__perc_id  target__perc_pos  \\\n",
       "1406  ENSGGOG00000003551          99.6330           99.7248   \n",
       "1407  ENSPPAG00000028735          99.7333           99.8222   \n",
       "1408  ENSPTRG00000010138          99.8222          100.0000   \n",
       "1409  ENSPPYG00000009265          98.7556           99.2000   \n",
       "1410  ENSNLEG00000008406          97.9556           98.8444   \n",
       "\n",
       "      target__protein_id      target__species  target__taxon_id  \\\n",
       "1406  ENSGGOP00000003487      gorilla_gorilla              9595   \n",
       "1407  ENSPPAP00000011050         pan_paniscus              9597   \n",
       "1408  ENSPTRP00000017225      pan_troglodytes              9598   \n",
       "1409  ENSPPYP00000010399         pongo_abelii              9601   \n",
       "1410  ENSNLEP00000010232  nomascus_leucogenys             61853   \n",
       "\n",
       "     taxonomy_level              type  \n",
       "1406      Homininae  ortholog_one2one  \n",
       "1407      Homininae  ortholog_one2one  \n",
       "1408      Homininae  ortholog_one2one  \n",
       "1409      Hominidae  ortholog_one2one  \n",
       "1410     Hominoidea  ortholog_one2one  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs = tfs_original.query('source__id in @tfs_random_subset.db_id')\n",
    "print(tfs.shape)\n",
    "tfs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Go to Quest for Orthologs fastas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read species metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000007062</td>\n",
       "      <td>7165</td>\n",
       "      <td>ANOGA</td>\n",
       "      <td>12553</td>\n",
       "      <td>971</td>\n",
       "      <td>13619</td>\n",
       "      <td>Anopheles gambiae (African malaria mosquito)</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "      <td>African malaria mosquito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UP000000798</td>\n",
       "      <td>224324</td>\n",
       "      <td>AQUAE</td>\n",
       "      <td>1553</td>\n",
       "      <td>0</td>\n",
       "      <td>1557</td>\n",
       "      <td>Aquifex aeolicus (strain VF5)</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>Aquifex aeolicus</td>\n",
       "      <td>strain VF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UP000006548</td>\n",
       "      <td>3702</td>\n",
       "      <td>ARATH</td>\n",
       "      <td>27475</td>\n",
       "      <td>14123</td>\n",
       "      <td>41920</td>\n",
       "      <td>Arabidopsis thaliana (Mouse-ear cress)</td>\n",
       "      <td>1496.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Mouse-ear cress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UP000001570</td>\n",
       "      <td>224308</td>\n",
       "      <td>BACSU</td>\n",
       "      <td>4260</td>\n",
       "      <td>7</td>\n",
       "      <td>4268</td>\n",
       "      <td>Bacillus subtilis (strain 168)</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain 168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UP000001414</td>\n",
       "      <td>226186</td>\n",
       "      <td>BACTN</td>\n",
       "      <td>4782</td>\n",
       "      <td>0</td>\n",
       "      <td>4823</td>\n",
       "      <td>Bacteroides thetaiotaomicron (strain ATCC 2914...</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>Bacteroides thetaiotaomicron</td>\n",
       "      <td>strain ATCC 29148 / DSM 2079 / NCTC 10582 / E5...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "0  UP000007062    7165  ANOGA                   12553   \n",
       "1  UP000000798  224324  AQUAE                    1553   \n",
       "2  UP000006548    3702  ARATH                   27475   \n",
       "3  UP000001570  224308  BACSU                    4260   \n",
       "4  UP000001414  226186  BACTN                    4782   \n",
       "\n",
       "   n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "0                          971       13619   \n",
       "1                            0        1557   \n",
       "2                        14123       41920   \n",
       "3                            7        4268   \n",
       "4                            0        4823   \n",
       "\n",
       "                                        species_name  \\\n",
       "0       Anopheles gambiae (African malaria mosquito)   \n",
       "1                      Aquifex aeolicus (strain VF5)   \n",
       "2             Arabidopsis thaliana (Mouse-ear cress)   \n",
       "3                     Bacillus subtilis (strain 168)   \n",
       "4  Bacteroides thetaiotaomicron (strain ATCC 2914...   \n",
       "\n",
       "   divergence_from_human_mya               scientific_name  \\\n",
       "0                      797.0             Anopheles gambiae   \n",
       "1                     4290.0              Aquifex aeolicus   \n",
       "2                     1496.0          Arabidopsis thaliana   \n",
       "3                     4290.0             Bacillus subtilis   \n",
       "4                     4290.0  Bacteroides thetaiotaomicron   \n",
       "\n",
       "                               common_name_or_strain  \n",
       "0                           African malaria mosquito  \n",
       "1                                         strain VF5  \n",
       "2                                    Mouse-ear cress  \n",
       "3                                         strain 168  \n",
       "4  strain ATCC 29148 / DSM 2079 / NCTC 10582 / E5...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_metadata = pd.read_csv(\"species_metadata.csv\")\n",
    "print(species_metadata.shape)\n",
    "species_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset to opisthokonts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000007062</td>\n",
       "      <td>7165</td>\n",
       "      <td>ANOGA</td>\n",
       "      <td>12553</td>\n",
       "      <td>971</td>\n",
       "      <td>13619</td>\n",
       "      <td>Anopheles gambiae (African malaria mosquito)</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "      <td>African malaria mosquito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UP000007241</td>\n",
       "      <td>684364</td>\n",
       "      <td>BATDJ</td>\n",
       "      <td>8610</td>\n",
       "      <td>0</td>\n",
       "      <td>8685</td>\n",
       "      <td>Batrachochytrium dendrobatidis (strain JAM81 /...</td>\n",
       "      <td>1105.0</td>\n",
       "      <td>Batrachochytrium dendrobatidis</td>\n",
       "      <td>strain JAM81 / FGSC 10211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UP000009136</td>\n",
       "      <td>9913</td>\n",
       "      <td>BOVIN</td>\n",
       "      <td>23774</td>\n",
       "      <td>14534</td>\n",
       "      <td>38438</td>\n",
       "      <td>Bos taurus (Bovine)</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Bovine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UP000001554</td>\n",
       "      <td>7739</td>\n",
       "      <td>BRAFL</td>\n",
       "      <td>28542</td>\n",
       "      <td>2</td>\n",
       "      <td>28608</td>\n",
       "      <td>Branchiostoma floridae (Florida lancelet) (Amp...</td>\n",
       "      <td>684.0</td>\n",
       "      <td>Branchiostoma floridae</td>\n",
       "      <td>Florida lancelet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UP000001940</td>\n",
       "      <td>6239</td>\n",
       "      <td>CAEEL</td>\n",
       "      <td>19986</td>\n",
       "      <td>8309</td>\n",
       "      <td>28507</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "0  UP000007062    7165  ANOGA                   12553   \n",
       "5  UP000007241  684364  BATDJ                    8610   \n",
       "6  UP000009136    9913  BOVIN                   23774   \n",
       "8  UP000001554    7739  BRAFL                   28542   \n",
       "9  UP000001940    6239  CAEEL                   19986   \n",
       "\n",
       "   n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "0                          971       13619   \n",
       "5                            0        8685   \n",
       "6                        14534       38438   \n",
       "8                            2       28608   \n",
       "9                         8309       28507   \n",
       "\n",
       "                                        species_name  \\\n",
       "0       Anopheles gambiae (African malaria mosquito)   \n",
       "5  Batrachochytrium dendrobatidis (strain JAM81 /...   \n",
       "6                                Bos taurus (Bovine)   \n",
       "8  Branchiostoma floridae (Florida lancelet) (Amp...   \n",
       "9                             Caenorhabditis elegans   \n",
       "\n",
       "   divergence_from_human_mya                 scientific_name  \\\n",
       "0                      797.0               Anopheles gambiae   \n",
       "5                     1105.0  Batrachochytrium dendrobatidis   \n",
       "6                       96.0                      Bos taurus   \n",
       "8                      684.0          Branchiostoma floridae   \n",
       "9                      797.0          Caenorhabditis elegans   \n",
       "\n",
       "       common_name_or_strain  \n",
       "0   African malaria mosquito  \n",
       "5  strain JAM81 / FGSC 10211  \n",
       "6                     Bovine  \n",
       "8           Florida lancelet  \n",
       "9                        NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Estimated opisthokonta divergence time from http://timetree.org/\n",
    "t = 1105\n",
    "opisthokonts = species_metadata.query('divergence_from_human_mya <= @t')\n",
    "print(opisthokonts.shape)\n",
    "opisthokonts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>UP000005640</td>\n",
       "      <td>9606</td>\n",
       "      <td>HUMAN</td>\n",
       "      <td>21146</td>\n",
       "      <td>74769</td>\n",
       "      <td>96332</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "32  UP000005640    9606  HUMAN                   21146   \n",
       "\n",
       "    n_additional_fasta_isoforms  n_gene2acc          species_name  \\\n",
       "32                        74769       96332  Homo sapiens (Human)   \n",
       "\n",
       "    divergence_from_human_mya scientific_name common_name_or_strain  \n",
       "32                        0.0    Homo sapiens                 Human  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opisthokonts.query('scientific_name == \"Homo sapiens\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Gene Accession file\n",
    "\n",
    "```\n",
    "Gene mapping files (*.gene2acc)\n",
    "===============================\n",
    "\n",
    "Column 1 is a unique gene symbol that is chosen with the following order of\n",
    "preference from the annotation found in:\n",
    "1) Model Organism Database (MOD)\n",
    "2) Ensembl or Ensembl Genomes database\n",
    "3) UniProt Ordered Locus Name (OLN)\n",
    "4) UniProt Open Reading Frame (ORF)\n",
    "5) UniProt Gene Name\n",
    "A dash symbol ('-') is used when the gene encoding a protein is unknown.\n",
    "\n",
    "Column 2 is the UniProtKB accession or isoform identifier for the given gene\n",
    "symbol. This column may have redundancy when two or more genes have identical\n",
    "translations.\n",
    "\n",
    "Column 3 is the gene symbol of the canonical accession used to represent the\n",
    "respective gene group and the first row of the sequence is the canonical one.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(96332, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maybe_ensembl_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>canonical_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A075B7B6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A075B714</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A075B713</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A075B712</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A075B711</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  maybe_ensembl_id  uniprot_id canonical_accession\n",
       "0              NaN  A0A075B7B6                 NaN\n",
       "1              NaN  A0A075B714                 NaN\n",
       "2              NaN  A0A075B713                 NaN\n",
       "3              NaN  A0A075B712                 NaN\n",
       "4              NaN  A0A075B711                 NaN"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def read_gene2acc(gene2acc, names=['maybe_ensembl_id', 'uniprot_id', 'canonical_accession']):\n",
    "    df = pd.read_csv(gene2acc, sep='\\t', header=None, na_values='-', names=names)\n",
    "    return df\n",
    "\n",
    "gene2acc = read_gene2acc('Eukaryota/UP000005640_9606.gene2acc')\n",
    "# gene2acc = pd.read_csv('Eukaryota/UP000005640_9606.gene2acc', sep='\\t', header=None, na_values='-', names=columns)\n",
    "print(gene2acc.shape)\n",
    "gene2acc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maybe_ensembl_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>canonical_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>ACOT7L</td>\n",
       "      <td>Q6ZUV0</td>\n",
       "      <td>ACOT7L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>BARGIN</td>\n",
       "      <td>Q6ZT62</td>\n",
       "      <td>BARGIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>BARGIN</td>\n",
       "      <td>Q6ZT62-2</td>\n",
       "      <td>BARGIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>BCE1</td>\n",
       "      <td>O60756</td>\n",
       "      <td>BCE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>C1orf140</td>\n",
       "      <td>Q5VVS0</td>\n",
       "      <td>C1orf140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96327</th>\n",
       "      <td>UNQ6190/PRO20217</td>\n",
       "      <td>Q6UXQ8</td>\n",
       "      <td>UNQ6190/PRO20217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96328</th>\n",
       "      <td>UNQ6493/PRO21345</td>\n",
       "      <td>Q6UXR8</td>\n",
       "      <td>UNQ6493/PRO21345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96329</th>\n",
       "      <td>UNQ6494/PRO21346</td>\n",
       "      <td>Q6UXR6</td>\n",
       "      <td>UNQ6494/PRO21346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96330</th>\n",
       "      <td>UNQ9165/PRO28630</td>\n",
       "      <td>Q6UXU0</td>\n",
       "      <td>UNQ9165/PRO28630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96331</th>\n",
       "      <td>UNQ9370/PRO34162</td>\n",
       "      <td>Q6UXP9</td>\n",
       "      <td>UNQ9370/PRO34162</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95732 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       maybe_ensembl_id uniprot_id canonical_accession\n",
       "600              ACOT7L     Q6ZUV0              ACOT7L\n",
       "601              BARGIN     Q6ZT62              BARGIN\n",
       "602              BARGIN   Q6ZT62-2              BARGIN\n",
       "603                BCE1     O60756                BCE1\n",
       "604            C1orf140     Q5VVS0            C1orf140\n",
       "...                 ...        ...                 ...\n",
       "96327  UNQ6190/PRO20217     Q6UXQ8    UNQ6190/PRO20217\n",
       "96328  UNQ6493/PRO21345     Q6UXR8    UNQ6493/PRO21345\n",
       "96329  UNQ6494/PRO21346     Q6UXR6    UNQ6494/PRO21346\n",
       "96330  UNQ9165/PRO28630     Q6UXU0    UNQ9165/PRO28630\n",
       "96331  UNQ9370/PRO34162     Q6UXP9    UNQ9370/PRO34162\n",
       "\n",
       "[95732 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2acc.dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read ID mapping file\n",
    "\n",
    "```\n",
    "Database mapping files (*.idmapping)\n",
    "====================================\n",
    "\n",
    "These files contain mappings from UniProtKB to other databases for each\n",
    "reference proteome.\n",
    "The format consists of three tab-separated columns:\n",
    "\n",
    "1. UniProtKB accession\n",
    "2. ID_type:\n",
    "   Database name as shown in UniProtKB cross-references and supported by the ID\n",
    "   mapping tool on the UniProt web site (http://www.uniprot.org/mapping)\n",
    "3. ID:\n",
    "   Identifier in the cross-referenced database.\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000007062</td>\n",
       "      <td>7165</td>\n",
       "      <td>ANOGA</td>\n",
       "      <td>12553</td>\n",
       "      <td>971</td>\n",
       "      <td>13619</td>\n",
       "      <td>Anopheles gambiae (African malaria mosquito)</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "      <td>African malaria mosquito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UP000007241</td>\n",
       "      <td>684364</td>\n",
       "      <td>BATDJ</td>\n",
       "      <td>8610</td>\n",
       "      <td>0</td>\n",
       "      <td>8685</td>\n",
       "      <td>Batrachochytrium dendrobatidis (strain JAM81 /...</td>\n",
       "      <td>1105.0</td>\n",
       "      <td>Batrachochytrium dendrobatidis</td>\n",
       "      <td>strain JAM81 / FGSC 10211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UP000009136</td>\n",
       "      <td>9913</td>\n",
       "      <td>BOVIN</td>\n",
       "      <td>23774</td>\n",
       "      <td>14534</td>\n",
       "      <td>38438</td>\n",
       "      <td>Bos taurus (Bovine)</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Bovine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UP000001554</td>\n",
       "      <td>7739</td>\n",
       "      <td>BRAFL</td>\n",
       "      <td>28542</td>\n",
       "      <td>2</td>\n",
       "      <td>28608</td>\n",
       "      <td>Branchiostoma floridae (Florida lancelet) (Amp...</td>\n",
       "      <td>684.0</td>\n",
       "      <td>Branchiostoma floridae</td>\n",
       "      <td>Florida lancelet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UP000001940</td>\n",
       "      <td>6239</td>\n",
       "      <td>CAEEL</td>\n",
       "      <td>19986</td>\n",
       "      <td>8309</td>\n",
       "      <td>28507</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "0  UP000007062    7165  ANOGA                   12553   \n",
       "5  UP000007241  684364  BATDJ                    8610   \n",
       "6  UP000009136    9913  BOVIN                   23774   \n",
       "8  UP000001554    7739  BRAFL                   28542   \n",
       "9  UP000001940    6239  CAEEL                   19986   \n",
       "\n",
       "   n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "0                          971       13619   \n",
       "5                            0        8685   \n",
       "6                        14534       38438   \n",
       "8                            2       28608   \n",
       "9                         8309       28507   \n",
       "\n",
       "                                        species_name  \\\n",
       "0       Anopheles gambiae (African malaria mosquito)   \n",
       "5  Batrachochytrium dendrobatidis (strain JAM81 /...   \n",
       "6                                Bos taurus (Bovine)   \n",
       "8  Branchiostoma floridae (Florida lancelet) (Amp...   \n",
       "9                             Caenorhabditis elegans   \n",
       "\n",
       "   divergence_from_human_mya                 scientific_name  \\\n",
       "0                      797.0               Anopheles gambiae   \n",
       "5                     1105.0  Batrachochytrium dendrobatidis   \n",
       "6                       96.0                      Bos taurus   \n",
       "8                      684.0          Branchiostoma floridae   \n",
       "9                      797.0          Caenorhabditis elegans   \n",
       "\n",
       "       common_name_or_strain  \n",
       "0   African malaria mosquito  \n",
       "5  strain JAM81 / FGSC 10211  \n",
       "6                     Bovine  \n",
       "8           Florida lancelet  \n",
       "9                        NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opisthokonts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>UP000000437</td>\n",
       "      <td>7955</td>\n",
       "      <td>DANRE</td>\n",
       "      <td>25939</td>\n",
       "      <td>21219</td>\n",
       "      <td>47517</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>435.0</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>Zebrafish</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "17  UP000000437    7955  DANRE                   25939   \n",
       "\n",
       "    n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "17                        21219       47517   \n",
       "\n",
       "                                   species_name  divergence_from_human_mya  \\\n",
       "17  Danio rerio (Zebrafish) (Brachydanio rerio)                      435.0   \n",
       "\n",
       "   scientific_name common_name_or_strain  \n",
       "17     Danio rerio             Zebrafish  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opisthokonts.query('proteome_id == \"UP000000437\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get ID Mapping for uniprot ids from ENSMBL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 1/48 [00:00<00:32,  1.44it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1036083, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  8%|▊         | 4/48 [00:01<00:18,  2.35it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(555421, 5)\n",
      "(149092, 5)\n",
      "(141961, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 12%|█▎        | 6/48 [00:02<00:19,  2.14it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1631036, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 17%|█▋        | 8/48 [00:02<00:16,  2.46it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(809657, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 21%|██        | 10/48 [00:03<00:13,  2.92it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(575653, 5)\n",
      "(295804, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 25%|██▌       | 12/48 [00:03<00:10,  3.59it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(124774, 5)\n",
      "(277055, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 27%|██▋       | 13/48 [00:03<00:07,  4.39it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(181143, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 33%|███▎      | 16/48 [00:04<00:06,  4.63it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(839553, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 40%|███▉      | 19/48 [00:04<00:05,  5.33it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(565651, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 42%|████▏     | 20/48 [00:05<00:06,  4.24it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(549426, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 46%|████▌     | 22/48 [00:05<00:05,  4.49it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(518273, 5)\n",
      "(205999, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 48%|████▊     | 23/48 [00:05<00:07,  3.56it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(765068, 5)\n",
      "(146229, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 54%|█████▍    | 26/48 [00:06<00:05,  4.11it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(594466, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 56%|█████▋    | 27/48 [00:07<00:08,  2.53it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1143307, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 60%|██████    | 29/48 [00:07<00:05,  3.32it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(424236, 5)\n",
      "(268306, 5)\n",
      "(149870, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 67%|██████▋   | 32/48 [00:08<00:04,  3.97it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(873723, 5)\n",
      "(214092, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 71%|███████   | 34/48 [00:09<00:05,  2.46it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2668934, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 79%|███████▉  | 38/48 [00:10<00:03,  3.33it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(340838, 5)\n",
      "(151313, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 83%|████████▎ | 40/48 [00:10<00:02,  3.92it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(315939, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 90%|████████▉ | 43/48 [00:10<00:01,  4.43it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(637633, 5)\n",
      "(281658, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 92%|█████████▏| 44/48 [00:11<00:00,  4.62it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(307893, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 94%|█████████▍| 45/48 [00:11<00:00,  3.24it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(773011, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 96%|█████████▌| 46/48 [00:12<00:00,  3.11it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(561122, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 48/48 [00:12<00:00,  3.90it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(390075, 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19464294, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>id_type</th>\n",
       "      <th>db_id</th>\n",
       "      <th>species_id</th>\n",
       "      <th>taxa_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>UniProtKB-ID</td>\n",
       "      <td>143BA_DANRE</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>Gene_Name</td>\n",
       "      <td>ywhaba</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>ywhab1</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>Gene_ORFName</td>\n",
       "      <td>wu:fb80c08</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>GI</td>\n",
       "      <td>82592598</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id       id_type        db_id   species_id      taxa_id\n",
       "0     Q5PRD0  UniProtKB-ID  143BA_DANRE  UP000000437  UP000000437\n",
       "1     Q5PRD0     Gene_Name       ywhaba  UP000000437  UP000000437\n",
       "2     Q5PRD0  Gene_Synonym       ywhab1  UP000000437  UP000000437\n",
       "3     Q5PRD0  Gene_ORFName   wu:fb80c08  UP000000437  UP000000437\n",
       "4     Q5PRD0            GI     82592598  UP000000437  UP000000437"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = []\n",
    "\n",
    "for filename in tqdm(sorted(iglob(\"Eukaryota/*.idmapping\"))):\n",
    "#     print(filename)\n",
    "    basename = os.path.basename(filename)\n",
    "    prefix = basename.split('.')[0]\n",
    "    species_id, taxa_id = prefix.split(\"_\")\n",
    "#     print(f\"{species_id=} {taxa_id=}\")\n",
    "    \n",
    "    if species_id in opisthokonts.proteome_id.values:\n",
    "\n",
    "        df = pd.read_csv(filename, sep='\\t', header=None, names=['uniprot_id', 'id_type', 'db_id'])\n",
    "        df['species_id'] = species_id\n",
    "        df['taxa_id'] = species_id\n",
    "\n",
    "        # Use only Ensembl data\n",
    "#         df = df.query('id_type == \"Ensembl\"')\n",
    "        print(df.shape)\n",
    "        dfs.append(df)\n",
    "    \n",
    "id_mapping = pd.concat(dfs, ignore_index=True)\n",
    "print(id_mapping.shape)\n",
    "id_mapping.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge id mapping with ensembl compara tfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>target__perc_pos</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340606</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000140265</td>\n",
       "      <td>79.8122</td>\n",
       "      <td>87.0892</td>\n",
       "      <td>ENSP00000380174</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPCIG00000018336</td>\n",
       "      <td>78.5219</td>\n",
       "      <td>85.6813</td>\n",
       "      <td>ENSPCIP00000024306</td>\n",
       "      <td>phascolarctos_cinereus</td>\n",
       "      <td>38626</td>\n",
       "      <td>Theria</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340607</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000140265</td>\n",
       "      <td>78.1690</td>\n",
       "      <td>87.2066</td>\n",
       "      <td>ENSP00000380174</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSSHAG00000012821</td>\n",
       "      <td>76.7281</td>\n",
       "      <td>85.5991</td>\n",
       "      <td>ENSSHAP00000015023</td>\n",
       "      <td>sarcophilus_harrisii</td>\n",
       "      <td>9305</td>\n",
       "      <td>Theria</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340608</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000140265</td>\n",
       "      <td>44.6009</td>\n",
       "      <td>48.7089</td>\n",
       "      <td>ENSP00000380174</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSMODG00000047609</td>\n",
       "      <td>82.2511</td>\n",
       "      <td>89.8268</td>\n",
       "      <td>ENSMODP00000047480</td>\n",
       "      <td>monodelphis_domestica</td>\n",
       "      <td>13616</td>\n",
       "      <td>Theria</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340609</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000140265</td>\n",
       "      <td>34.7418</td>\n",
       "      <td>41.6667</td>\n",
       "      <td>ENSP00000380174</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSOANG00000010234</td>\n",
       "      <td>55.6391</td>\n",
       "      <td>66.7293</td>\n",
       "      <td>ENSOANP00000016232</td>\n",
       "      <td>ornithorhynchus_anatinus</td>\n",
       "      <td>9258</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340610</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000140265</td>\n",
       "      <td>12.0892</td>\n",
       "      <td>22.8873</td>\n",
       "      <td>ENSP00000380174</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSGALG00000046739</td>\n",
       "      <td>17.5768</td>\n",
       "      <td>33.2765</td>\n",
       "      <td>ENSGALP00000069629</td>\n",
       "      <td>gallus_gallus</td>\n",
       "      <td>9031</td>\n",
       "      <td>Amniota</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dn_ds     method_link_type       source__id  source__perc_id  \\\n",
       "340606    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000140265          79.8122   \n",
       "340607    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000140265          78.1690   \n",
       "340608    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000140265          44.6009   \n",
       "340609    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000140265          34.7418   \n",
       "340610    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000140265          12.0892   \n",
       "\n",
       "        source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "340606           87.0892    ENSP00000380174    homo_sapiens              9606   \n",
       "340607           87.2066    ENSP00000380174    homo_sapiens              9606   \n",
       "340608           48.7089    ENSP00000380174    homo_sapiens              9606   \n",
       "340609           41.6667    ENSP00000380174    homo_sapiens              9606   \n",
       "340610           22.8873    ENSP00000380174    homo_sapiens              9606   \n",
       "\n",
       "                target__id  target__perc_id  target__perc_pos  \\\n",
       "340606  ENSPCIG00000018336          78.5219           85.6813   \n",
       "340607  ENSSHAG00000012821          76.7281           85.5991   \n",
       "340608  ENSMODG00000047609          82.2511           89.8268   \n",
       "340609  ENSOANG00000010234          55.6391           66.7293   \n",
       "340610  ENSGALG00000046739          17.5768           33.2765   \n",
       "\n",
       "        target__protein_id           target__species  target__taxon_id  \\\n",
       "340606  ENSPCIP00000024306    phascolarctos_cinereus             38626   \n",
       "340607  ENSSHAP00000015023      sarcophilus_harrisii              9305   \n",
       "340608  ENSMODP00000047480     monodelphis_domestica             13616   \n",
       "340609  ENSOANP00000016232  ornithorhynchus_anatinus              9258   \n",
       "340610  ENSGALP00000069629             gallus_gallus              9031   \n",
       "\n",
       "       taxonomy_level               type  \n",
       "340606         Theria   ortholog_one2one  \n",
       "340607         Theria   ortholog_one2one  \n",
       "340608         Theria  ortholog_one2many  \n",
       "340609       Mammalia   ortholog_one2one  \n",
       "340610        Amniota   ortholog_one2one  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target__uniprot_id</th>\n",
       "      <th>target__id_type</th>\n",
       "      <th>target__db_id</th>\n",
       "      <th>target__species_id</th>\n",
       "      <th>target__taxa_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>UniProtKB-ID</td>\n",
       "      <td>143BA_DANRE</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>Gene_Name</td>\n",
       "      <td>ywhaba</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>ywhab1</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>Gene_ORFName</td>\n",
       "      <td>wu:fb80c08</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q5PRD0</td>\n",
       "      <td>GI</td>\n",
       "      <td>82592598</td>\n",
       "      <td>UP000000437</td>\n",
       "      <td>UP000000437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target__uniprot_id target__id_type target__db_id target__species_id  \\\n",
       "0             Q5PRD0    UniProtKB-ID   143BA_DANRE        UP000000437   \n",
       "1             Q5PRD0       Gene_Name        ywhaba        UP000000437   \n",
       "2             Q5PRD0    Gene_Synonym        ywhab1        UP000000437   \n",
       "3             Q5PRD0    Gene_ORFName    wu:fb80c08        UP000000437   \n",
       "4             Q5PRD0              GI      82592598        UP000000437   \n",
       "\n",
       "  target__taxa_id  \n",
       "0     UP000000437  \n",
       "1     UP000000437  \n",
       "2     UP000000437  \n",
       "3     UP000000437  \n",
       "4     UP000000437  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapping_for_merging = id_mapping.copy()\n",
    "id_mapping_for_merging.columns = \"target__\" + id_mapping_for_merging.columns\n",
    "id_mapping_for_merging.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20586, 16)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3105, 21)\n",
      "CPU times: user 11.3 s, sys: 460 ms, total: 11.7 s\n",
      "Wall time: 11.7 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>...</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "      <th>target__uniprot_id</th>\n",
       "      <th>target__id_type</th>\n",
       "      <th>target__db_id</th>\n",
       "      <th>target__species_id</th>\n",
       "      <th>target__taxa_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.15000</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>96.0212</td>\n",
       "      <td>96.1096</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>99.6330</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>gorilla_gorilla</td>\n",
       "      <td>9595</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>G3QLW5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>UP000001519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.08163</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>99.2927</td>\n",
       "      <td>99.4695</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPTRG00000010138</td>\n",
       "      <td>99.8222</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSPTRP00000017225</td>\n",
       "      <td>pan_troglodytes</td>\n",
       "      <td>9598</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>H2QER5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSPTRP00000017225</td>\n",
       "      <td>UP000002277</td>\n",
       "      <td>UP000002277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.34381</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>58.0018</td>\n",
       "      <td>71.0875</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSRNOG00000053843</td>\n",
       "      <td>57.1429</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSRNOP00000073516</td>\n",
       "      <td>rattus_norvegicus</td>\n",
       "      <td>10116</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>A0A0G2K5Q5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSRNOP00000073516</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>UP000002494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.13859</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>78.7798</td>\n",
       "      <td>85.9416</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSRNOG00000053370</td>\n",
       "      <td>73.8194</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSRNOP00000068569</td>\n",
       "      <td>rattus_norvegicus</td>\n",
       "      <td>10116</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>A0A0G2JT49</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSRNOP00000068569</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>UP000002494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.15195</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>80.1061</td>\n",
       "      <td>87.0911</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSMUSG00000053950</td>\n",
       "      <td>77.7682</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSP00000068560</td>\n",
       "      <td>mus_musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>Q8CHC8</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSMUSP00000068560</td>\n",
       "      <td>UP000000589</td>\n",
       "      <td>UP000000589</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dn_ds     method_link_type       source__id  source__perc_id  \\\n",
       "0  0.15000  ENSEMBL_ORTHOLOGUES  ENSG00000101544          96.0212   \n",
       "1  0.08163  ENSEMBL_ORTHOLOGUES  ENSG00000101544          99.2927   \n",
       "2  0.34381  ENSEMBL_ORTHOLOGUES  ENSG00000101544          58.0018   \n",
       "3  0.13859  ENSEMBL_ORTHOLOGUES  ENSG00000101544          78.7798   \n",
       "4  0.15195  ENSEMBL_ORTHOLOGUES  ENSG00000101544          80.1061   \n",
       "\n",
       "   source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "0           96.1096    ENSP00000262198    homo_sapiens              9606   \n",
       "1           99.4695    ENSP00000262198    homo_sapiens              9606   \n",
       "2           71.0875    ENSP00000262198    homo_sapiens              9606   \n",
       "3           85.9416    ENSP00000262198    homo_sapiens              9606   \n",
       "4           87.0911    ENSP00000262198    homo_sapiens              9606   \n",
       "\n",
       "           target__id  target__perc_id  ...  target__protein_id  \\\n",
       "0  ENSGGOG00000003551          99.6330  ...  ENSGGOP00000003487   \n",
       "1  ENSPTRG00000010138          99.8222  ...  ENSPTRP00000017225   \n",
       "2  ENSRNOG00000053843          57.1429  ...  ENSRNOP00000073516   \n",
       "3  ENSRNOG00000053370          73.8194  ...  ENSRNOP00000068569   \n",
       "4  ENSMUSG00000053950          77.7682  ...  ENSMUSP00000068560   \n",
       "\n",
       "     target__species target__taxon_id    taxonomy_level               type  \\\n",
       "0    gorilla_gorilla             9595         Homininae   ortholog_one2one   \n",
       "1    pan_troglodytes             9598         Homininae   ortholog_one2one   \n",
       "2  rattus_norvegicus            10116  Euarchontoglires  ortholog_one2many   \n",
       "3  rattus_norvegicus            10116  Euarchontoglires  ortholog_one2many   \n",
       "4       mus_musculus            10090  Euarchontoglires   ortholog_one2one   \n",
       "\n",
       "  target__uniprot_id target__id_type       target__db_id target__species_id  \\\n",
       "0             G3QLW5     Ensembl_PRO  ENSGGOP00000003487        UP000001519   \n",
       "1             H2QER5     Ensembl_PRO  ENSPTRP00000017225        UP000002277   \n",
       "2         A0A0G2K5Q5     Ensembl_PRO  ENSRNOP00000073516        UP000002494   \n",
       "3         A0A0G2JT49     Ensembl_PRO  ENSRNOP00000068569        UP000002494   \n",
       "4             Q8CHC8     Ensembl_PRO  ENSMUSP00000068560        UP000000589   \n",
       "\n",
       "  target__taxa_id  \n",
       "0     UP000001519  \n",
       "1     UP000002277  \n",
       "2     UP000002494  \n",
       "3     UP000002494  \n",
       "4     UP000000589  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "tfs_uniprot_merge_proteins = tfs.merge(id_mapping_for_merging, left_on='target__protein_id', right_on='target__db_id')\n",
    "print(tfs_uniprot_merge_proteins.shape)\n",
    "tfs_uniprot_merge_proteins.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13846, 21)\n",
      "CPU times: user 10.7 s, sys: 418 ms, total: 11.1 s\n",
      "Wall time: 11.1 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>...</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "      <th>target__uniprot_id</th>\n",
       "      <th>target__id_type</th>\n",
       "      <th>target__db_id</th>\n",
       "      <th>target__species_id</th>\n",
       "      <th>target__taxa_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.15000</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>96.0212</td>\n",
       "      <td>96.1096</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>99.6330</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>gorilla_gorilla</td>\n",
       "      <td>9595</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>A0A2I2ZMG9</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>UP000001519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.15000</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>96.0212</td>\n",
       "      <td>96.1096</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>99.6330</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>gorilla_gorilla</td>\n",
       "      <td>9595</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>G3QLW5</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>UP000001519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.08163</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>99.2927</td>\n",
       "      <td>99.4695</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPTRG00000010138</td>\n",
       "      <td>99.8222</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSPTRP00000017225</td>\n",
       "      <td>pan_troglodytes</td>\n",
       "      <td>9598</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>H2QER5</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSPTRG00000010138</td>\n",
       "      <td>UP000002277</td>\n",
       "      <td>UP000002277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.34381</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>58.0018</td>\n",
       "      <td>71.0875</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSRNOG00000053843</td>\n",
       "      <td>57.1429</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSRNOP00000073516</td>\n",
       "      <td>rattus_norvegicus</td>\n",
       "      <td>10116</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>A0A0G2K5Q5</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSRNOG00000053843</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>UP000002494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.13859</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>78.7798</td>\n",
       "      <td>85.9416</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSRNOG00000053370</td>\n",
       "      <td>73.8194</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSRNOP00000068569</td>\n",
       "      <td>rattus_norvegicus</td>\n",
       "      <td>10116</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>A0A0G2JT49</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSRNOG00000053370</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>UP000002494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dn_ds     method_link_type       source__id  source__perc_id  \\\n",
       "0  0.15000  ENSEMBL_ORTHOLOGUES  ENSG00000101544          96.0212   \n",
       "1  0.15000  ENSEMBL_ORTHOLOGUES  ENSG00000101544          96.0212   \n",
       "2  0.08163  ENSEMBL_ORTHOLOGUES  ENSG00000101544          99.2927   \n",
       "3  0.34381  ENSEMBL_ORTHOLOGUES  ENSG00000101544          58.0018   \n",
       "4  0.13859  ENSEMBL_ORTHOLOGUES  ENSG00000101544          78.7798   \n",
       "\n",
       "   source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "0           96.1096    ENSP00000262198    homo_sapiens              9606   \n",
       "1           96.1096    ENSP00000262198    homo_sapiens              9606   \n",
       "2           99.4695    ENSP00000262198    homo_sapiens              9606   \n",
       "3           71.0875    ENSP00000262198    homo_sapiens              9606   \n",
       "4           85.9416    ENSP00000262198    homo_sapiens              9606   \n",
       "\n",
       "           target__id  target__perc_id  ...  target__protein_id  \\\n",
       "0  ENSGGOG00000003551          99.6330  ...  ENSGGOP00000003487   \n",
       "1  ENSGGOG00000003551          99.6330  ...  ENSGGOP00000003487   \n",
       "2  ENSPTRG00000010138          99.8222  ...  ENSPTRP00000017225   \n",
       "3  ENSRNOG00000053843          57.1429  ...  ENSRNOP00000073516   \n",
       "4  ENSRNOG00000053370          73.8194  ...  ENSRNOP00000068569   \n",
       "\n",
       "     target__species target__taxon_id    taxonomy_level               type  \\\n",
       "0    gorilla_gorilla             9595         Homininae   ortholog_one2one   \n",
       "1    gorilla_gorilla             9595         Homininae   ortholog_one2one   \n",
       "2    pan_troglodytes             9598         Homininae   ortholog_one2one   \n",
       "3  rattus_norvegicus            10116  Euarchontoglires  ortholog_one2many   \n",
       "4  rattus_norvegicus            10116  Euarchontoglires  ortholog_one2many   \n",
       "\n",
       "  target__uniprot_id target__id_type       target__db_id target__species_id  \\\n",
       "0         A0A2I2ZMG9         Ensembl  ENSGGOG00000003551        UP000001519   \n",
       "1             G3QLW5         Ensembl  ENSGGOG00000003551        UP000001519   \n",
       "2             H2QER5         Ensembl  ENSPTRG00000010138        UP000002277   \n",
       "3         A0A0G2K5Q5         Ensembl  ENSRNOG00000053843        UP000002494   \n",
       "4         A0A0G2JT49         Ensembl  ENSRNOG00000053370        UP000002494   \n",
       "\n",
       "  target__taxa_id  \n",
       "0     UP000001519  \n",
       "1     UP000001519  \n",
       "2     UP000002277  \n",
       "3     UP000002494  \n",
       "4     UP000002494  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "tfs_uniprot_merge_genes = tfs.merge(id_mapping_for_merging, left_on='target__id', right_on='target__db_id')\n",
    "print(tfs_uniprot_merge_genes.shape)\n",
    "tfs_uniprot_merge_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add human uniprot ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2668934, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source__uniprot_id</th>\n",
       "      <th>source__id_type</th>\n",
       "      <th>source__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P62258-2</td>\n",
       "      <td>UniParc</td>\n",
       "      <td>UPI00001E6021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P62258-2</td>\n",
       "      <td>Ensembl_TRS</td>\n",
       "      <td>ENST00000571732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P62258-2</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000461762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P62258-2</td>\n",
       "      <td>Ensembl_TRS</td>\n",
       "      <td>ENST00000616643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P62258-2</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000481059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source__uniprot_id source__id_type    source__db_id\n",
       "0           P62258-2         UniParc    UPI00001E6021\n",
       "1           P62258-2     Ensembl_TRS  ENST00000571732\n",
       "2           P62258-2     Ensembl_PRO  ENSP00000461762\n",
       "3           P62258-2     Ensembl_TRS  ENST00000616643\n",
       "4           P62258-2     Ensembl_PRO  ENSP00000481059"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_id_mapping = pd.read_csv('Eukaryota/UP000005640_9606.idmapping', sep='\\t', header=None, names=['uniprot_id', 'id_type', 'db_id'])\n",
    "human_id_mapping.columns = 'source__' + human_id_mapping.columns\n",
    "print(human_id_mapping.shape)\n",
    "human_id_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2671985, 24)\n",
      "CPU times: user 2.03 s, sys: 340 ms, total: 2.37 s\n",
      "Wall time: 2.36 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>human__id</th>\n",
       "      <th>human__perc_id</th>\n",
       "      <th>human__perc_pos</th>\n",
       "      <th>human__protein_id</th>\n",
       "      <th>human__species</th>\n",
       "      <th>human__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "      <th>target__uniprot_id</th>\n",
       "      <th>target__id_type</th>\n",
       "      <th>target__db_id</th>\n",
       "      <th>target__species_id</th>\n",
       "      <th>target__taxa_id</th>\n",
       "      <th>human__uniprot_id</th>\n",
       "      <th>human__id_type</th>\n",
       "      <th>human__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.15000</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>96.0212</td>\n",
       "      <td>96.1096</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>99.6330</td>\n",
       "      <td>...</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>G3QLW5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.08163</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>99.2927</td>\n",
       "      <td>99.4695</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSPTRG00000010138</td>\n",
       "      <td>99.8222</td>\n",
       "      <td>...</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>H2QER5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSPTRP00000017225</td>\n",
       "      <td>UP000002277</td>\n",
       "      <td>UP000002277</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.15195</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>80.1061</td>\n",
       "      <td>87.0911</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSMUSG00000053950</td>\n",
       "      <td>77.7682</td>\n",
       "      <td>...</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>Q8CHC8</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSMUSP00000068560</td>\n",
       "      <td>UP000000589</td>\n",
       "      <td>UP000000589</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.13911</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>86.2069</td>\n",
       "      <td>90.8046</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSCAFG00000000007</td>\n",
       "      <td>85.9788</td>\n",
       "      <td>...</td>\n",
       "      <td>Boreoeutheria</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>F1PJN7</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSCAFP00000000007</td>\n",
       "      <td>UP000002254</td>\n",
       "      <td>UP000002254</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.13181</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>86.0301</td>\n",
       "      <td>91.1583</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSBTAG00000005916</td>\n",
       "      <td>84.5352</td>\n",
       "      <td>...</td>\n",
       "      <td>Boreoeutheria</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>A0A3S5ZPP3</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSBTAP00000007769</td>\n",
       "      <td>UP000009136</td>\n",
       "      <td>UP000009136</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dn_ds     method_link_type        human__id  human__perc_id  \\\n",
       "0  0.15000  ENSEMBL_ORTHOLOGUES  ENSG00000101544         96.0212   \n",
       "1  0.08163  ENSEMBL_ORTHOLOGUES  ENSG00000101544         99.2927   \n",
       "4  0.15195  ENSEMBL_ORTHOLOGUES  ENSG00000101544         80.1061   \n",
       "5  0.13911  ENSEMBL_ORTHOLOGUES  ENSG00000101544         86.2069   \n",
       "6  0.13181  ENSEMBL_ORTHOLOGUES  ENSG00000101544         86.0301   \n",
       "\n",
       "   human__perc_pos human__protein_id human__species  human__taxon_id  \\\n",
       "0          96.1096   ENSP00000262198   homo_sapiens           9606.0   \n",
       "1          99.4695   ENSP00000262198   homo_sapiens           9606.0   \n",
       "4          87.0911   ENSP00000262198   homo_sapiens           9606.0   \n",
       "5          90.8046   ENSP00000262198   homo_sapiens           9606.0   \n",
       "6          91.1583   ENSP00000262198   homo_sapiens           9606.0   \n",
       "\n",
       "           target__id  target__perc_id  ...    taxonomy_level  \\\n",
       "0  ENSGGOG00000003551          99.6330  ...         Homininae   \n",
       "1  ENSPTRG00000010138          99.8222  ...         Homininae   \n",
       "4  ENSMUSG00000053950          77.7682  ...  Euarchontoglires   \n",
       "5  ENSCAFG00000000007          85.9788  ...     Boreoeutheria   \n",
       "6  ENSBTAG00000005916          84.5352  ...     Boreoeutheria   \n",
       "\n",
       "               type target__uniprot_id  target__id_type       target__db_id  \\\n",
       "0  ortholog_one2one             G3QLW5      Ensembl_PRO  ENSGGOP00000003487   \n",
       "1  ortholog_one2one             H2QER5      Ensembl_PRO  ENSPTRP00000017225   \n",
       "4  ortholog_one2one             Q8CHC8      Ensembl_PRO  ENSMUSP00000068560   \n",
       "5  ortholog_one2one             F1PJN7      Ensembl_PRO  ENSCAFP00000000007   \n",
       "6  ortholog_one2one         A0A3S5ZPP3      Ensembl_PRO  ENSBTAP00000007769   \n",
       "\n",
       "  target__species_id target__taxa_id human__uniprot_id human__id_type  \\\n",
       "0        UP000001519     UP000001519            Q6IQ32    Ensembl_PRO   \n",
       "1        UP000002277     UP000002277            Q6IQ32    Ensembl_PRO   \n",
       "4        UP000000589     UP000000589            Q6IQ32    Ensembl_PRO   \n",
       "5        UP000002254     UP000002254            Q6IQ32    Ensembl_PRO   \n",
       "6        UP000009136     UP000009136            Q6IQ32    Ensembl_PRO   \n",
       "\n",
       "      human__db_id  \n",
       "0  ENSP00000262198  \n",
       "1  ENSP00000262198  \n",
       "4  ENSP00000262198  \n",
       "5  ENSP00000262198  \n",
       "6  ENSP00000262198  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time \n",
    "tfs_uniprot_merge_proteins_with_human = tfs_uniprot_merge_proteins.merge(\n",
    "    human_id_mapping, left_on='source__protein_id', right_on='source__db_id', how='outer')\n",
    "tfs_uniprot_merge_proteins_with_human.columns = tfs_uniprot_merge_proteins_with_human.columns.str.replace(\"source__\", 'human__')\n",
    "print(tfs_uniprot_merge_proteins_with_human.shape)\n",
    "tfs_uniprot_merge_proteins_with_human.query('type == \"ortholog_one2one\"').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "other_paralog             1893\n",
       "ortholog_one2one           707\n",
       "ortholog_many2many         233\n",
       "ortholog_one2many          184\n",
       "within_species_paralog      86\n",
       "alt_allele                   2\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs_uniprot_merge_proteins_with_human.type.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write merged TFs to disk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 27.5 s, sys: 53 ms, total: 27.6 s\n",
      "Wall time: 27.8 s\n",
      "CPU times: user 2.36 s, sys: 134 ms, total: 2.5 s\n",
      "Wall time: 2.64 s\n"
     ]
    }
   ],
   "source": [
    "%time tfs_uniprot_merge_proteins_with_human.to_csv('opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot_random_subset100.csv.gz', index=False)\n",
    "%time tfs_uniprot_merge_proteins_with_human.to_parquet('opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot_random_subset100.parquet', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make set variable for quick membership evalution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_orthologs = set(tfs_uniprot_merge_proteins_with_human.target__uniprot_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prove that the set `tf_orthologs` is faster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26.5 ns ± 0.024 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit 'Q7Z761' in tf_orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.62 µs ± 35.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit 'Q7Z761' in tfs_uniprot_merge_proteins_with_human.target__uniprot_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Yep, sets are 3 orders of magnitude faster!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read non-human proteins and subset if they are an ortholog of a TF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make outdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34mhuman-protein-fastas\u001b[0m/\n",
      "\u001b[01;34mhuman-protein-fastas-concatenated\u001b[0m/\n",
      "\u001b[01;34mhuman-transcription-factor-fastas\u001b[0m/\n",
      "\u001b[01;34mhuman-transcription-factor-fastas-random-subset100\u001b[0m/\n",
      "\u001b[01;34mhuman-visual-transduction-fastas\u001b[0m/\n",
      "\u001b[01;34mnot-human-protein-fastas\u001b[0m/\n",
      "\u001b[01;34mnot-human-transcription-factor-fastas\u001b[0m/\n",
      "\u001b[01;34mnot-human-visual-transduction-fastas\u001b[0m/\n",
      "\u001b[01;34mprotein-fastas\u001b[0m/\n",
      "UP000000437_7955_additional.fasta\n",
      "UP000000437_7955_DNA.fasta\n",
      "UP000000437_7955_DNA.miss\n",
      "UP000000437_7955.fasta\n",
      "UP000000437_7955_gcoord.xml\n",
      "UP000000437_7955.gene2acc\n",
      "UP000000437_7955.idmapping\n",
      "UP000000437_7955.xml\n",
      "UP000000539_9031_additional.fasta\n",
      "UP000000539_9031_DNA.fasta\n",
      "UP000000539_9031_DNA.miss\n",
      "UP000000539_9031.fasta\n",
      "UP000000539_9031_gcoord.xml\n",
      "UP000000539_9031.gene2acc\n",
      "UP000000539_9031.idmapping\n",
      "UP000000539_9031.xml\n",
      "UP000000542_5664_DNA.fasta\n",
      "UP000000542_5664.fasta\n",
      "UP000000542_5664.gene2acc\n",
      "UP000000542_5664.idmapping\n",
      "UP000000542_5664.xml\n",
      "UP000000559_237561_additional.fasta\n",
      "UP000000559_237561_DNA.fasta\n",
      "UP000000559_237561_DNA.miss\n",
      "UP000000559_237561.fasta\n",
      "UP000000559_237561.gene2acc\n",
      "UP000000559_237561.idmapping\n",
      "UP000000559_237561.xml\n",
      "UP000000561_237631_additional.fasta\n",
      "UP000000561_237631_DNA.fasta\n",
      "UP000000561_237631.fasta\n",
      "UP000000561_237631.gene2acc\n",
      "UP000000561_237631.idmapping\n",
      "UP000000561_237631.xml\n",
      "UP000000589_10090_additional.fasta\n",
      "UP000000589_10090_DNA.fasta\n",
      "UP000000589_10090_DNA.miss\n",
      "UP000000589_10090.fasta\n",
      "UP000000589_10090_gcoord.xml\n",
      "UP000000589_10090.gene2acc\n",
      "UP000000589_10090.idmapping\n",
      "UP000000589_10090.xml\n",
      "UP000000600_5888_DNA.fasta\n",
      "UP000000600_5888.fasta\n",
      "UP000000600_5888.gene2acc\n",
      "UP000000600_5888.idmapping\n",
      "UP000000600_5888.xml\n",
      "UP000000803_7227_additional.fasta\n",
      "UP000000803_7227_DNA.fasta\n",
      "UP000000803_7227.fasta\n",
      "UP000000803_7227.gene2acc\n",
      "UP000000803_7227.idmapping\n",
      "UP000000803_7227.xml\n",
      "UP000001038_8090_additional.fasta\n",
      "UP000001038_8090_DNA.fasta\n",
      "UP000001038_8090_DNA.miss\n",
      "UP000001038_8090.fasta\n",
      "UP000001038_8090_gcoord.xml\n",
      "UP000001038_8090.gene2acc\n",
      "UP000001038_8090.idmapping\n",
      "UP000001038_8090.xml\n",
      "UP000001055_321614_DNA.fasta\n",
      "UP000001055_321614.fasta\n",
      "UP000001055_321614.gene2acc\n",
      "UP000001055_321614.idmapping\n",
      "UP000001055_321614.xml\n",
      "UP000001300_284591_additional.fasta\n",
      "UP000001300_284591_DNA.fasta\n",
      "UP000001300_284591_DNA.miss\n",
      "UP000001300_284591.fasta\n",
      "UP000001300_284591.gene2acc\n",
      "UP000001300_284591.idmapping\n",
      "UP000001300_284591.xml\n",
      "UP000001312_665079_DNA.fasta\n",
      "UP000001312_665079.fasta\n",
      "UP000001312_665079.gene2acc\n",
      "UP000001312_665079.idmapping\n",
      "UP000001312_665079.xml\n",
      "UP000001357_81824_DNA.fasta\n",
      "UP000001357_81824.fasta\n",
      "UP000001357_81824.gene2acc\n",
      "UP000001357_81824.idmapping\n",
      "UP000001357_81824.xml\n",
      "UP000001449_35128_additional.fasta\n",
      "UP000001449_35128_DNA.fasta\n",
      "UP000001449_35128.fasta\n",
      "UP000001449_35128.gene2acc\n",
      "UP000001449_35128.idmapping\n",
      "UP000001449_35128.xml\n",
      "UP000001450_36329_additional.fasta\n",
      "UP000001450_36329_DNA.fasta\n",
      "UP000001450_36329_DNA.miss\n",
      "UP000001450_36329.fasta\n",
      "UP000001450_36329.gene2acc\n",
      "UP000001450_36329.idmapping\n",
      "UP000001450_36329.xml\n",
      "UP000001519_9595_additional.fasta\n",
      "UP000001519_9595_DNA.fasta\n",
      "UP000001519_9595_DNA.miss\n",
      "UP000001519_9595.fasta\n",
      "UP000001519_9595_gcoord.xml\n",
      "UP000001519_9595.gene2acc\n",
      "UP000001519_9595.idmapping\n",
      "UP000001519_9595.xml\n",
      "UP000001542_5722_DNA.fasta\n",
      "UP000001542_5722.fasta\n",
      "UP000001542_5722.gene2acc\n",
      "UP000001542_5722.idmapping\n",
      "UP000001542_5722.xml\n",
      "UP000001548_184922_DNA.fasta\n",
      "UP000001548_184922.fasta\n",
      "UP000001548_184922.gene2acc\n",
      "UP000001548_184922.idmapping\n",
      "UP000001548_184922.xml\n",
      "UP000001554_7739_additional.fasta\n",
      "UP000001554_7739_DNA.fasta\n",
      "UP000001554_7739_DNA.miss\n",
      "UP000001554_7739.fasta\n",
      "UP000001554_7739.gene2acc\n",
      "UP000001554_7739.idmapping\n",
      "UP000001554_7739.xml\n",
      "UP000001555_6945_DNA.fasta\n",
      "UP000001555_6945.fasta\n",
      "UP000001555_6945_gcoord.xml\n",
      "UP000001555_6945.gene2acc\n",
      "UP000001555_6945.idmapping\n",
      "UP000001555_6945.xml\n",
      "UP000001593_45351_additional.fasta\n",
      "UP000001593_45351_DNA.fasta\n",
      "UP000001593_45351.fasta\n",
      "UP000001593_45351.gene2acc\n",
      "UP000001593_45351.idmapping\n",
      "UP000001593_45351.xml\n",
      "UP000001805_367110_additional.fasta\n",
      "UP000001805_367110_DNA.fasta\n",
      "UP000001805_367110.fasta\n",
      "UP000001805_367110.gene2acc\n",
      "UP000001805_367110.idmapping\n",
      "UP000001805_367110.xml\n",
      "UP000001940_6239_additional.fasta\n",
      "UP000001940_6239_DNA.fasta\n",
      "UP000001940_6239_DNA.miss\n",
      "UP000001940_6239.fasta\n",
      "UP000001940_6239.gene2acc\n",
      "UP000001940_6239.idmapping\n",
      "UP000001940_6239.xml\n",
      "UP000002149_214684_additional.fasta\n",
      "UP000002149_214684_DNA.fasta\n",
      "UP000002149_214684_DNA.miss\n",
      "UP000002149_214684.fasta\n",
      "UP000002149_214684.gene2acc\n",
      "UP000002149_214684.idmapping\n",
      "UP000002149_214684.xml\n",
      "UP000002195_44689_additional.fasta\n",
      "UP000002195_44689_DNA.fasta\n",
      "UP000002195_44689_DNA.miss\n",
      "UP000002195_44689.fasta\n",
      "UP000002195_44689.gene2acc\n",
      "UP000002195_44689.idmapping\n",
      "UP000002195_44689.xml\n",
      "UP000002254_9615_additional.fasta\n",
      "UP000002254_9615_DNA.fasta\n",
      "UP000002254_9615_DNA.miss\n",
      "UP000002254_9615.fasta\n",
      "UP000002254_9615_gcoord.xml\n",
      "UP000002254_9615.gene2acc\n",
      "UP000002254_9615.idmapping\n",
      "UP000002254_9615.xml\n",
      "UP000002277_9598_additional.fasta\n",
      "UP000002277_9598_DNA.fasta\n",
      "UP000002277_9598_DNA.miss\n",
      "UP000002277_9598.fasta\n",
      "UP000002277_9598_gcoord.xml\n",
      "UP000002277_9598.gene2acc\n",
      "UP000002277_9598.idmapping\n",
      "UP000002277_9598.xml\n",
      "UP000002280_13616_additional.fasta\n",
      "UP000002280_13616_DNA.fasta\n",
      "UP000002280_13616_DNA.miss\n",
      "UP000002280_13616.fasta\n",
      "UP000002280_13616_gcoord.xml\n",
      "UP000002280_13616.gene2acc\n",
      "UP000002280_13616.idmapping\n",
      "UP000002280_13616.xml\n",
      "UP000002311_559292_additional.fasta\n",
      "UP000002311_559292_DNA.fasta\n",
      "UP000002311_559292_DNA.miss\n",
      "UP000002311_559292.fasta\n",
      "UP000002311_559292_gcoord.xml\n",
      "UP000002311_559292.gene2acc\n",
      "UP000002311_559292.idmapping\n",
      "UP000002311_559292.xml\n",
      "UP000002485_284812_additional.fasta\n",
      "UP000002485_284812_DNA.fasta\n",
      "UP000002485_284812_DNA.miss\n",
      "UP000002485_284812.fasta\n",
      "UP000002485_284812.gene2acc\n",
      "UP000002485_284812.idmapping\n",
      "UP000002485_284812.xml\n",
      "UP000002494_10116_additional.fasta\n",
      "UP000002494_10116_DNA.fasta\n",
      "UP000002494_10116_DNA.miss\n",
      "UP000002494_10116.fasta\n",
      "UP000002494_10116_gcoord.xml\n",
      "UP000002494_10116.gene2acc\n",
      "UP000002494_10116.idmapping\n",
      "UP000002494_10116.xml\n",
      "UP000002530_330879_DNA.fasta\n",
      "UP000002530_330879.fasta\n",
      "UP000002530_330879.gene2acc\n",
      "UP000002530_330879.idmapping\n",
      "UP000002530_330879.xml\n",
      "UP000005238_164328_DNA.fasta\n",
      "UP000005238_164328.fasta\n",
      "UP000005238_164328_gcoord.xml\n",
      "UP000005238_164328.gene2acc\n",
      "UP000005238_164328.idmapping\n",
      "UP000005238_164328.xml\n",
      "UP000005640_9606_additional.fasta\n",
      "UP000005640_9606_DNA.fasta\n",
      "UP000005640_9606_DNA.miss\n",
      "UP000005640_9606.fasta\n",
      "UP000005640_9606_gcoord.miss\n",
      "UP000005640_9606_gcoord.xml\n",
      "UP000005640_9606.gene2acc\n",
      "UP000005640_9606.idmapping\n",
      "UP000005640_9606.xml\n",
      "UP000006548_3702_additional.fasta\n",
      "UP000006548_3702_DNA.fasta\n",
      "UP000006548_3702_DNA.miss\n",
      "UP000006548_3702.fasta\n",
      "UP000006548_3702_gcoord.xml\n",
      "UP000006548_3702.gene2acc\n",
      "UP000006548_3702.idmapping\n",
      "UP000006548_3702.xml\n",
      "UP000006727_3218_additional.fasta\n",
      "UP000006727_3218_DNA.fasta\n",
      "UP000006727_3218.fasta\n",
      "UP000006727_3218.gene2acc\n",
      "UP000006727_3218.idmapping\n",
      "UP000006727_3218.xml\n",
      "UP000006906_3055_additional.fasta\n",
      "UP000006906_3055_DNA.fasta\n",
      "UP000006906_3055_DNA.miss\n",
      "UP000006906_3055.fasta\n",
      "UP000006906_3055.gene2acc\n",
      "UP000006906_3055.idmapping\n",
      "UP000006906_3055.xml\n",
      "UP000007062_7165_additional.fasta\n",
      "UP000007062_7165_DNA.fasta\n",
      "UP000007062_7165_DNA.miss\n",
      "UP000007062_7165.fasta\n",
      "UP000007062_7165_gcoord.xml\n",
      "UP000007062_7165.gene2acc\n",
      "UP000007062_7165.idmapping\n",
      "UP000007062_7165.xml\n",
      "UP000007241_684364_DNA.fasta\n",
      "UP000007241_684364.fasta\n",
      "UP000007241_684364.gene2acc\n",
      "UP000007241_684364.idmapping\n",
      "UP000007241_684364.xml\n",
      "UP000007266_7070_additional.fasta\n",
      "UP000007266_7070_DNA.fasta\n",
      "UP000007266_7070.fasta\n",
      "UP000007266_7070.gene2acc\n",
      "UP000007266_7070.idmapping\n",
      "UP000007266_7070.xml\n",
      "UP000007305_4577_additional.fasta\n",
      "UP000007305_4577_DNA.fasta\n",
      "UP000007305_4577_DNA.miss\n",
      "UP000007305_4577.fasta\n",
      "UP000007305_4577_gcoord.xml\n",
      "UP000007305_4577.gene2acc\n",
      "UP000007305_4577.idmapping\n",
      "UP000007305_4577.xml\n",
      "UP000008143_8364_additional.fasta\n",
      "UP000008143_8364_DNA.fasta\n",
      "UP000008143_8364_DNA.miss\n",
      "UP000008143_8364.fasta\n",
      "UP000008143_8364_gcoord.xml\n",
      "UP000008143_8364.gene2acc\n",
      "UP000008143_8364.idmapping\n",
      "UP000008143_8364.xml\n",
      "UP000008144_7719_additional.fasta\n",
      "UP000008144_7719_DNA.fasta\n",
      "UP000008144_7719.fasta\n",
      "UP000008144_7719_gcoord.xml\n",
      "UP000008144_7719.gene2acc\n",
      "UP000008144_7719.idmapping\n",
      "UP000008144_7719.xml\n",
      "UP000008783_418459_additional.fasta\n",
      "UP000008783_418459_DNA.fasta\n",
      "UP000008783_418459.fasta\n",
      "UP000008783_418459.gene2acc\n",
      "UP000008783_418459.idmapping\n",
      "UP000008783_418459.xml\n",
      "UP000009136_9913_additional.fasta\n",
      "UP000009136_9913_DNA.fasta\n",
      "UP000009136_9913_DNA.miss\n",
      "UP000009136_9913.fasta\n",
      "UP000009136_9913_gcoord.xml\n",
      "UP000009136_9913.gene2acc\n",
      "UP000009136_9913.idmapping\n",
      "UP000009136_9913.xml\n",
      "UP000015101_6412_DNA.fasta\n",
      "UP000015101_6412.fasta\n",
      "UP000015101_6412_gcoord.xml\n",
      "UP000015101_6412.gene2acc\n",
      "UP000015101_6412.idmapping\n",
      "UP000015101_6412.xml\n",
      "UP000018468_7918_additional.fasta\n",
      "UP000018468_7918_DNA.fasta\n",
      "UP000018468_7918.fasta\n",
      "UP000018468_7918_gcoord.xml\n",
      "UP000018468_7918.gene2acc\n",
      "UP000018468_7918.idmapping\n",
      "UP000018468_7918.xml\n",
      "UP000059680_39947_additional.fasta\n",
      "UP000059680_39947_DNA.fasta\n",
      "UP000059680_39947_DNA.miss\n",
      "UP000059680_39947.fasta\n",
      "UP000059680_39947.gene2acc\n",
      "UP000059680_39947.idmapping\n",
      "UP000059680_39947.xml\n"
     ]
    }
   ],
   "source": [
    "ls Eukaryota"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How much compute is this?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of human transcription factor proteins in the quest for orthologs database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_human_qfo_tfs = 5749"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>human__id</th>\n",
       "      <th>human__perc_id</th>\n",
       "      <th>human__perc_pos</th>\n",
       "      <th>human__protein_id</th>\n",
       "      <th>human__species</th>\n",
       "      <th>human__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "      <th>target__uniprot_id</th>\n",
       "      <th>target__id_type</th>\n",
       "      <th>target__db_id</th>\n",
       "      <th>target__species_id</th>\n",
       "      <th>target__taxa_id</th>\n",
       "      <th>human__uniprot_id</th>\n",
       "      <th>human__id_type</th>\n",
       "      <th>human__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.15000</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>96.0212</td>\n",
       "      <td>96.1096</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSGGOG00000003551</td>\n",
       "      <td>99.6330</td>\n",
       "      <td>...</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>G3QLW5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSGGOP00000003487</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>UP000001519</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.08163</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>99.2927</td>\n",
       "      <td>99.4695</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSPTRG00000010138</td>\n",
       "      <td>99.8222</td>\n",
       "      <td>...</td>\n",
       "      <td>Homininae</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>H2QER5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSPTRP00000017225</td>\n",
       "      <td>UP000002277</td>\n",
       "      <td>UP000002277</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.34381</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>58.0018</td>\n",
       "      <td>71.0875</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSRNOG00000053843</td>\n",
       "      <td>57.1429</td>\n",
       "      <td>...</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>A0A0G2K5Q5</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSRNOP00000073516</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.13859</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>78.7798</td>\n",
       "      <td>85.9416</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSRNOG00000053370</td>\n",
       "      <td>73.8194</td>\n",
       "      <td>...</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>A0A0G2JT49</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSRNOP00000068569</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>UP000002494</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.15195</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000101544</td>\n",
       "      <td>80.1061</td>\n",
       "      <td>87.0911</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>ENSMUSG00000053950</td>\n",
       "      <td>77.7682</td>\n",
       "      <td>...</td>\n",
       "      <td>Euarchontoglires</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>Q8CHC8</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSMUSP00000068560</td>\n",
       "      <td>UP000000589</td>\n",
       "      <td>UP000000589</td>\n",
       "      <td>Q6IQ32</td>\n",
       "      <td>Ensembl_PRO</td>\n",
       "      <td>ENSP00000262198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dn_ds     method_link_type        human__id  human__perc_id  \\\n",
       "0  0.15000  ENSEMBL_ORTHOLOGUES  ENSG00000101544         96.0212   \n",
       "1  0.08163  ENSEMBL_ORTHOLOGUES  ENSG00000101544         99.2927   \n",
       "2  0.34381  ENSEMBL_ORTHOLOGUES  ENSG00000101544         58.0018   \n",
       "3  0.13859  ENSEMBL_ORTHOLOGUES  ENSG00000101544         78.7798   \n",
       "4  0.15195  ENSEMBL_ORTHOLOGUES  ENSG00000101544         80.1061   \n",
       "\n",
       "   human__perc_pos human__protein_id human__species  human__taxon_id  \\\n",
       "0          96.1096   ENSP00000262198   homo_sapiens           9606.0   \n",
       "1          99.4695   ENSP00000262198   homo_sapiens           9606.0   \n",
       "2          71.0875   ENSP00000262198   homo_sapiens           9606.0   \n",
       "3          85.9416   ENSP00000262198   homo_sapiens           9606.0   \n",
       "4          87.0911   ENSP00000262198   homo_sapiens           9606.0   \n",
       "\n",
       "           target__id  target__perc_id  ...    taxonomy_level  \\\n",
       "0  ENSGGOG00000003551          99.6330  ...         Homininae   \n",
       "1  ENSPTRG00000010138          99.8222  ...         Homininae   \n",
       "2  ENSRNOG00000053843          57.1429  ...  Euarchontoglires   \n",
       "3  ENSRNOG00000053370          73.8194  ...  Euarchontoglires   \n",
       "4  ENSMUSG00000053950          77.7682  ...  Euarchontoglires   \n",
       "\n",
       "                type target__uniprot_id  target__id_type       target__db_id  \\\n",
       "0   ortholog_one2one             G3QLW5      Ensembl_PRO  ENSGGOP00000003487   \n",
       "1   ortholog_one2one             H2QER5      Ensembl_PRO  ENSPTRP00000017225   \n",
       "2  ortholog_one2many         A0A0G2K5Q5      Ensembl_PRO  ENSRNOP00000073516   \n",
       "3  ortholog_one2many         A0A0G2JT49      Ensembl_PRO  ENSRNOP00000068569   \n",
       "4   ortholog_one2one             Q8CHC8      Ensembl_PRO  ENSMUSP00000068560   \n",
       "\n",
       "  target__species_id target__taxa_id human__uniprot_id human__id_type  \\\n",
       "0        UP000001519     UP000001519            Q6IQ32    Ensembl_PRO   \n",
       "1        UP000002277     UP000002277            Q6IQ32    Ensembl_PRO   \n",
       "2        UP000002494     UP000002494            Q6IQ32    Ensembl_PRO   \n",
       "3        UP000002494     UP000002494            Q6IQ32    Ensembl_PRO   \n",
       "4        UP000000589     UP000000589            Q6IQ32    Ensembl_PRO   \n",
       "\n",
       "      human__db_id  \n",
       "0  ENSP00000262198  \n",
       "1  ENSP00000262198  \n",
       "2  ENSP00000262198  \n",
       "3  ENSP00000262198  \n",
       "4  ENSP00000262198  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs_uniprot_merge_proteins_with_human.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1045"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs_uniprot_merge_proteins_with_human.query('target__species != \"homo_sapiens\"').target__uniprot_id.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_not_human_qfo_tfs = 1045"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0012841666666665"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_human_qfo_tfs * n_not_human_qfo_tfs * 0.0006 / 60 / 60"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in protein fastas with screed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘Eukaryota/not-human-transcription-factor-fastas-random-subset100/’: File exists\n",
      "Eukaryota/not-human-protein-fastas/UP000000437_7955_additional.fasta\n",
      "\tlen(tf_records): 13\n",
      "Eukaryota/not-human-protein-fastas/UP000008143_8364.fasta\n",
      "\tlen(tf_records): 12\n",
      "Eukaryota/not-human-protein-fastas/UP000001038_8090.fasta\n",
      "\tlen(tf_records): 85\n",
      "Eukaryota/not-human-protein-fastas/UP000002280_13616.fasta\n",
      "\tlen(tf_records): 52\n",
      "Eukaryota/not-human-protein-fastas/UP000000539_9031_additional.fasta\n",
      "\tlen(tf_records): 3\n",
      "Eukaryota/not-human-protein-fastas/UP000000803_7227_additional.fasta\n",
      "\tlen(tf_records): 20\n",
      "Eukaryota/not-human-protein-fastas/UP000018468_7918.fasta\n",
      "\tlen(tf_records): 68\n",
      "Eukaryota/not-human-protein-fastas/UP000002254_9615.fasta\n",
      "\tlen(tf_records): 51\n",
      "Eukaryota/not-human-protein-fastas/UP000001940_6239.fasta\n",
      "\tlen(tf_records): 7\n",
      "Eukaryota/not-human-protein-fastas/UP000000437_7955.fasta\n",
      "\tlen(tf_records): 76\n",
      "Eukaryota/not-human-protein-fastas/UP000002311_559292.fasta\n",
      "\tlen(tf_records): 8\n",
      "Eukaryota/not-human-protein-fastas/UP000000589_10090.fasta\n",
      "\tlen(tf_records): 59\n",
      "Eukaryota/not-human-protein-fastas/UP000002254_9615_additional.fasta\n",
      "\tlen(tf_records): 7\n",
      "Eukaryota/not-human-protein-fastas/UP000000803_7227.fasta\n",
      "\tlen(tf_records): 59\n",
      "Eukaryota/not-human-protein-fastas/UP000001519_9595_additional.fasta\n",
      "\tlen(tf_records): 1\n",
      "Eukaryota/not-human-protein-fastas/UP000002494_10116.fasta\n",
      "\tlen(tf_records): 69\n",
      "Eukaryota/not-human-protein-fastas/UP000001519_9595.fasta\n",
      "\tlen(tf_records): 87\n",
      "Eukaryota/not-human-protein-fastas/UP000008143_8364_additional.fasta\n",
      "\tlen(tf_records): 6\n",
      "Eukaryota/not-human-protein-fastas/UP000001940_6239_additional.fasta\n",
      "\tlen(tf_records): 1\n",
      "Eukaryota/not-human-protein-fastas/UP000000539_9031.fasta\n",
      "\tlen(tf_records): 68\n",
      "Eukaryota/not-human-protein-fastas/UP000000589_10090_additional.fasta\n",
      "\tlen(tf_records): 25\n",
      "Eukaryota/not-human-protein-fastas/UP000009136_9913.fasta\n",
      "\tlen(tf_records): 83\n",
      "Eukaryota/not-human-protein-fastas/UP000002277_9598.fasta\n",
      "\tlen(tf_records): 96\n",
      "Eukaryota/not-human-protein-fastas/UP000008144_7719.fasta\n",
      "\tlen(tf_records): 64\n",
      "Eukaryota/not-human-protein-fastas/UP000002494_10116_additional.fasta\n",
      "\tlen(tf_records): 19\n",
      "Eukaryota/not-human-protein-fastas/UP000002277_9598_additional.fasta\n",
      "\tlen(tf_records): 1\n",
      "Eukaryota/not-human-protein-fastas/UP000009136_9913_additional.fasta\n",
      "\tlen(tf_records): 5\n"
     ]
    }
   ],
   "source": [
    "not_human_outdir = 'Eukaryota/not-human-transcription-factor-fastas-random-subset100/'\n",
    "! mkdir $not_human_outdir\n",
    "\n",
    "\n",
    "for filename in iglob('Eukaryota/not-human-protein-fastas/*.fasta'):\n",
    "    tf_records = []\n",
    "\n",
    "    basename = os.path.basename(filename)\n",
    "    with screed.open(filename) as records:\n",
    "        for record in records:\n",
    "            name = record['name']\n",
    "            record_id = name.split()[0]\n",
    "            uniprot_id = record_id.split('|')[1]\n",
    "            if uniprot_id in tf_orthologs:\n",
    "                tf_records.append(record)\n",
    "                \n",
    "    if len(tf_records) > 0:\n",
    "        print(filename)\n",
    "        print(f\"\\tlen(tf_records): {len(tf_records)}\")\n",
    "        with open(f'{not_human_outdir}/{basename}', 'w') as f:\n",
    "            for record in tf_records:\n",
    "                f.write(\">{name}\\n{sequence}\\n\".format(**record))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Script to run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file qfo_human_vs_opisthokont_tfs.sh\n",
    "#!/bin/bash\n",
    "OUTDIR=$HOME/data_sm/kmer-hashing/quest-for-orthologs/analysis/2019/transcription-factors-random-subset100/\n",
    "mkdir -p $OUTDIR/intermediates\n",
    "cd $OUTDIR/intermediates\n",
    "\n",
    "PARQUET=$OUTDIR/qfo-eukaryota-tfs-protein.parquet\n",
    "\n",
    "EUKARYOTA=/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019/Eukaryota\n",
    "HUMAN=$EUKARYOTA/human-transcription-factor-fastas-random-subset100/human_transcription_factor_proteins.fasta\n",
    "NOT_HUMAN=$EUKARYOTA/not-human-transcription-factor-fastas-random-subset100/\n",
    "\n",
    "conda activate khtools--encodings--compare-cli\n",
    "\n",
    "\n",
    "time khtools compare-kmers \\\n",
    "    --processes 120 \\\n",
    "    --ksize-min 3 \\\n",
    "    --ksize-max 45 \\\n",
    "    --parquet $PARQUET \\\n",
    "    --intermediate-parquet \\\n",
    "    --fastas2 $HUMAN \\\n",
    "    $NOT_HUMAN/* | tee khtools_compare-kmers.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Time estimation\n",
    "\n",
    "taking ~1000 seconds per non-human sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_not_human_qfo_tfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_not_human_qfo_tfs * 1000 / 120 / 60 / 60 / 24"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, so this will take ~4.7 days to compute running on `lrrr`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:khtools--compare-kmers--encodings]",
   "language": "python",
   "name": "conda-env-khtools--compare-kmers--encodings-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
