{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import iglob\n",
    "import os\n",
    "from time import sleep\n",
    "\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "import pandas as pd\n",
    "import screed\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019\n"
     ]
    }
   ],
   "source": [
    "cd /mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 2.6G\n",
      "drwxr-xr-x 5 olga root 4.0K Jan 10 15:34 \u001b[0m\u001b[01;34m.\u001b[0m/\n",
      "drwxr-xr-x 3 olga root 4.0K Dec 25 17:48 \u001b[01;34m..\u001b[0m/\n",
      "drwxr-xr-x 5 olga czb  4.0K Dec 26 19:44 \u001b[01;34mArchaea\u001b[0m/\n",
      "drwxr-xr-x 5 olga czb   16K Dec 26 19:44 \u001b[01;34mBacteria\u001b[0m/\n",
      "drwxr-xr-x 9 olga czb   32K Jan 10 15:35 \u001b[01;34mEukaryota\u001b[0m/\n",
      "-rw-r--r-- 1 olga czb  754K Jan 10 07:50 human_transcription_factors_with_uniprot_ids.csv\n",
      "-rw-r--r-- 1 olga czb   68K Jan 10 07:50 \u001b[01;31mhuman_transcription_factors_with_uniprot_ids.csv.gz\u001b[0m\n",
      "-rw-r--r-- 1 olga czb  133K Jan 10 07:50 human_transcription_factors_with_uniprot_ids.parquet\n",
      "-rw-r--r-- 1 olga czb  4.5K Jan 10 15:34 human_visual_transduction_with_uniprot_ids.csv\n",
      "-rw-r--r-- 1 olga czb  1.2K Jan 10 15:34 \u001b[01;31mhuman_visual_transduction_with_uniprot_ids.csv.gz\u001b[0m\n",
      "-rw-r--r-- 1 olga czb  3.7K Jan 10 15:34 human_visual_transduction_with_uniprot_ids.parquet\n",
      "-rw-r--r-- 1 olga czb   64M Jan 10 07:43 opisthokont_not_human_transcription_factors_ensembl_compara.csv\n",
      "-rw-r--r-- 1 olga czb  1.8M Jan 10 08:02 \u001b[01;31mopisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.csv.gz\u001b[0m\u001b[K\n",
      "-rw-r--r-- 1 olga czb  2.2M Jan 10 08:02 opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.parquet\n",
      "-rw-r--r-- 1 olga czb   12M Jan 10 07:43 opisthokont_not_human_transcription_factors_ensembl_compara.parquet\n",
      "-rw-r--r-- 1 olga czb   661 Jan  8 17:33 qfo_human_vs_opisthokont_tfs.sh\n",
      "-rw-r--r-- 1 olga czb  2.6G Dec 25 18:46 \u001b[01;31mQfO_release_2019_04.tar.gz\u001b[0m\n",
      "-rw-r--r-- 1 olga czb   18K May 10  2019 README\n",
      "-rw-r--r-- 1 olga czb   12K Jan  8 07:47 species_metadata.csv\n"
     ]
    }
   ],
   "source": [
    "ls -lha"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read human visual transduction genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3053, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source__uniprot_id</th>\n",
       "      <th>source__id_type</th>\n",
       "      <th>source__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P29973</td>\n",
       "      <td>UniProtKB-ID</td>\n",
       "      <td>CNGA1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Name</td>\n",
       "      <td>CNGA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>CNCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>CNCG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P29973</td>\n",
       "      <td>GI</td>\n",
       "      <td>2506302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source__uniprot_id source__id_type source__db_id\n",
       "0             P29973    UniProtKB-ID   CNGA1_HUMAN\n",
       "1             P29973       Gene_Name         CNGA1\n",
       "2             P29973    Gene_Synonym          CNCG\n",
       "3             P29973    Gene_Synonym         CNCG1\n",
       "4             P29973              GI       2506302"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual = pd.read_csv(\"human_visual_transduction_with_uniprot_ids.csv\")\n",
    "print(visual.shape)\n",
    "visual.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Go to Quest for Orthologs fastas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019\n"
     ]
    }
   ],
   "source": [
    "cd /home/olga/data_sm/kmer-hashing/quest-for-orthologs/data/2019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 2.6G\n",
      "drwxr-xr-x 5 olga root 4.0K Jan 10 15:34 \u001b[0m\u001b[01;34m.\u001b[0m/\n",
      "drwxr-xr-x 3 olga root 4.0K Dec 25 17:48 \u001b[01;34m..\u001b[0m/\n",
      "drwxr-xr-x 5 olga czb  4.0K Dec 26 19:44 \u001b[01;34mArchaea\u001b[0m/\n",
      "drwxr-xr-x 5 olga czb   16K Dec 26 19:44 \u001b[01;34mBacteria\u001b[0m/\n",
      "drwxr-xr-x 9 olga czb   32K Jan 10 15:35 \u001b[01;34mEukaryota\u001b[0m/\n",
      "-rw-r--r-- 1 olga czb  754K Jan 10 07:50 human_transcription_factors_with_uniprot_ids.csv\n",
      "-rw-r--r-- 1 olga czb   68K Jan 10 07:50 \u001b[01;31mhuman_transcription_factors_with_uniprot_ids.csv.gz\u001b[0m\n",
      "-rw-r--r-- 1 olga czb  133K Jan 10 07:50 human_transcription_factors_with_uniprot_ids.parquet\n",
      "-rw-r--r-- 1 olga czb   76K Jan 15 10:57 human_visual_transduction_with_uniprot_ids.csv\n",
      "-rw-r--r-- 1 olga czb   19K Jan 15 10:57 \u001b[01;31mhuman_visual_transduction_with_uniprot_ids.csv.gz\u001b[0m\n",
      "-rw-r--r-- 1 olga czb   28K Jan 15 10:57 human_visual_transduction_with_uniprot_ids.parquet\n",
      "-rw-r--r-- 1 olga czb   64M Jan 10 07:43 opisthokont_not_human_transcription_factors_ensembl_compara.csv\n",
      "-rw-r--r-- 1 olga czb  1.8M Jan 10 08:02 \u001b[01;31mopisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.csv.gz\u001b[0m\u001b[K\n",
      "-rw-r--r-- 1 olga czb  2.2M Jan 10 08:02 opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.parquet\n",
      "-rw-r--r-- 1 olga czb   12M Jan 10 07:43 opisthokont_not_human_transcription_factors_ensembl_compara.parquet\n",
      "-rw-r--r-- 1 olga czb   661 Jan  8 17:33 qfo_human_vs_opisthokont_tfs.sh\n",
      "-rw-r--r-- 1 olga czb  2.6G Dec 25 18:46 \u001b[01;31mQfO_release_2019_04.tar.gz\u001b[0m\n",
      "-rw-r--r-- 1 olga czb   18K May 10  2019 README\n",
      "-rw-r--r-- 1 olga czb   12K Jan  8 07:47 species_metadata.csv\n"
     ]
    }
   ],
   "source": [
    "ls -lha"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read species metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000007062</td>\n",
       "      <td>7165</td>\n",
       "      <td>ANOGA</td>\n",
       "      <td>12553</td>\n",
       "      <td>971</td>\n",
       "      <td>13619</td>\n",
       "      <td>Anopheles gambiae (African malaria mosquito)</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "      <td>African malaria mosquito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UP000000798</td>\n",
       "      <td>224324</td>\n",
       "      <td>AQUAE</td>\n",
       "      <td>1553</td>\n",
       "      <td>0</td>\n",
       "      <td>1557</td>\n",
       "      <td>Aquifex aeolicus (strain VF5)</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>Aquifex aeolicus</td>\n",
       "      <td>strain VF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UP000006548</td>\n",
       "      <td>3702</td>\n",
       "      <td>ARATH</td>\n",
       "      <td>27475</td>\n",
       "      <td>14123</td>\n",
       "      <td>41920</td>\n",
       "      <td>Arabidopsis thaliana (Mouse-ear cress)</td>\n",
       "      <td>1496.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Mouse-ear cress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UP000001570</td>\n",
       "      <td>224308</td>\n",
       "      <td>BACSU</td>\n",
       "      <td>4260</td>\n",
       "      <td>7</td>\n",
       "      <td>4268</td>\n",
       "      <td>Bacillus subtilis (strain 168)</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain 168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UP000001414</td>\n",
       "      <td>226186</td>\n",
       "      <td>BACTN</td>\n",
       "      <td>4782</td>\n",
       "      <td>0</td>\n",
       "      <td>4823</td>\n",
       "      <td>Bacteroides thetaiotaomicron (strain ATCC 2914...</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>Bacteroides thetaiotaomicron</td>\n",
       "      <td>strain ATCC 29148 / DSM 2079 / NCTC 10582 / E5...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "0  UP000007062    7165  ANOGA                   12553   \n",
       "1  UP000000798  224324  AQUAE                    1553   \n",
       "2  UP000006548    3702  ARATH                   27475   \n",
       "3  UP000001570  224308  BACSU                    4260   \n",
       "4  UP000001414  226186  BACTN                    4782   \n",
       "\n",
       "   n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "0                          971       13619   \n",
       "1                            0        1557   \n",
       "2                        14123       41920   \n",
       "3                            7        4268   \n",
       "4                            0        4823   \n",
       "\n",
       "                                        species_name  \\\n",
       "0       Anopheles gambiae (African malaria mosquito)   \n",
       "1                      Aquifex aeolicus (strain VF5)   \n",
       "2             Arabidopsis thaliana (Mouse-ear cress)   \n",
       "3                     Bacillus subtilis (strain 168)   \n",
       "4  Bacteroides thetaiotaomicron (strain ATCC 2914...   \n",
       "\n",
       "   divergence_from_human_mya               scientific_name  \\\n",
       "0                      797.0             Anopheles gambiae   \n",
       "1                     4290.0              Aquifex aeolicus   \n",
       "2                     1496.0          Arabidopsis thaliana   \n",
       "3                     4290.0             Bacillus subtilis   \n",
       "4                     4290.0  Bacteroides thetaiotaomicron   \n",
       "\n",
       "                               common_name_or_strain  \n",
       "0                           African malaria mosquito  \n",
       "1                                         strain VF5  \n",
       "2                                    Mouse-ear cress  \n",
       "3                                         strain 168  \n",
       "4  strain ATCC 29148 / DSM 2079 / NCTC 10582 / E5...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_metadata = pd.read_csv(\"species_metadata.csv\")\n",
    "print(species_metadata.shape)\n",
    "species_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset to opisthokonts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000007062</td>\n",
       "      <td>7165</td>\n",
       "      <td>ANOGA</td>\n",
       "      <td>12553</td>\n",
       "      <td>971</td>\n",
       "      <td>13619</td>\n",
       "      <td>Anopheles gambiae (African malaria mosquito)</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "      <td>African malaria mosquito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UP000007241</td>\n",
       "      <td>684364</td>\n",
       "      <td>BATDJ</td>\n",
       "      <td>8610</td>\n",
       "      <td>0</td>\n",
       "      <td>8685</td>\n",
       "      <td>Batrachochytrium dendrobatidis (strain JAM81 /...</td>\n",
       "      <td>1105.0</td>\n",
       "      <td>Batrachochytrium dendrobatidis</td>\n",
       "      <td>strain JAM81 / FGSC 10211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UP000009136</td>\n",
       "      <td>9913</td>\n",
       "      <td>BOVIN</td>\n",
       "      <td>23774</td>\n",
       "      <td>14534</td>\n",
       "      <td>38438</td>\n",
       "      <td>Bos taurus (Bovine)</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Bovine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UP000001554</td>\n",
       "      <td>7739</td>\n",
       "      <td>BRAFL</td>\n",
       "      <td>28542</td>\n",
       "      <td>2</td>\n",
       "      <td>28608</td>\n",
       "      <td>Branchiostoma floridae (Florida lancelet) (Amp...</td>\n",
       "      <td>684.0</td>\n",
       "      <td>Branchiostoma floridae</td>\n",
       "      <td>Florida lancelet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UP000001940</td>\n",
       "      <td>6239</td>\n",
       "      <td>CAEEL</td>\n",
       "      <td>19986</td>\n",
       "      <td>8309</td>\n",
       "      <td>28507</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "0  UP000007062    7165  ANOGA                   12553   \n",
       "5  UP000007241  684364  BATDJ                    8610   \n",
       "6  UP000009136    9913  BOVIN                   23774   \n",
       "8  UP000001554    7739  BRAFL                   28542   \n",
       "9  UP000001940    6239  CAEEL                   19986   \n",
       "\n",
       "   n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "0                          971       13619   \n",
       "5                            0        8685   \n",
       "6                        14534       38438   \n",
       "8                            2       28608   \n",
       "9                         8309       28507   \n",
       "\n",
       "                                        species_name  \\\n",
       "0       Anopheles gambiae (African malaria mosquito)   \n",
       "5  Batrachochytrium dendrobatidis (strain JAM81 /...   \n",
       "6                                Bos taurus (Bovine)   \n",
       "8  Branchiostoma floridae (Florida lancelet) (Amp...   \n",
       "9                             Caenorhabditis elegans   \n",
       "\n",
       "   divergence_from_human_mya                 scientific_name  \\\n",
       "0                      797.0               Anopheles gambiae   \n",
       "5                     1105.0  Batrachochytrium dendrobatidis   \n",
       "6                       96.0                      Bos taurus   \n",
       "8                      684.0          Branchiostoma floridae   \n",
       "9                      797.0          Caenorhabditis elegans   \n",
       "\n",
       "       common_name_or_strain  \n",
       "0   African malaria mosquito  \n",
       "5  strain JAM81 / FGSC 10211  \n",
       "6                     Bovine  \n",
       "8           Florida lancelet  \n",
       "9                        NaN  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Estimated opisthokonta divergence time from http://timetree.org/\n",
    "t = 1105\n",
    "opisthokonts = species_metadata.query('divergence_from_human_mya <= @t')\n",
    "print(opisthokonts.shape)\n",
    "opisthokonts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>UP000005640</td>\n",
       "      <td>9606</td>\n",
       "      <td>HUMAN</td>\n",
       "      <td>21146</td>\n",
       "      <td>74769</td>\n",
       "      <td>96332</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "32  UP000005640    9606  HUMAN                   21146   \n",
       "\n",
       "    n_additional_fasta_isoforms  n_gene2acc          species_name  \\\n",
       "32                        74769       96332  Homo sapiens (Human)   \n",
       "\n",
       "    divergence_from_human_mya scientific_name common_name_or_strain  \n",
       "32                        0.0    Homo sapiens                 Human  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opisthokonts.query('scientific_name == \"Homo sapiens\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(34, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteome_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>oscode</th>\n",
       "      <th>n_main_fasta_canonical</th>\n",
       "      <th>n_additional_fasta_isoforms</th>\n",
       "      <th>n_gene2acc</th>\n",
       "      <th>species_name</th>\n",
       "      <th>divergence_from_human_mya</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name_or_strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000007062</td>\n",
       "      <td>7165</td>\n",
       "      <td>ANOGA</td>\n",
       "      <td>12553</td>\n",
       "      <td>971</td>\n",
       "      <td>13619</td>\n",
       "      <td>Anopheles gambiae (African malaria mosquito)</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "      <td>African malaria mosquito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UP000007241</td>\n",
       "      <td>684364</td>\n",
       "      <td>BATDJ</td>\n",
       "      <td>8610</td>\n",
       "      <td>0</td>\n",
       "      <td>8685</td>\n",
       "      <td>Batrachochytrium dendrobatidis (strain JAM81 /...</td>\n",
       "      <td>1105.0</td>\n",
       "      <td>Batrachochytrium dendrobatidis</td>\n",
       "      <td>strain JAM81 / FGSC 10211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UP000009136</td>\n",
       "      <td>9913</td>\n",
       "      <td>BOVIN</td>\n",
       "      <td>23774</td>\n",
       "      <td>14534</td>\n",
       "      <td>38438</td>\n",
       "      <td>Bos taurus (Bovine)</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Bovine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UP000001554</td>\n",
       "      <td>7739</td>\n",
       "      <td>BRAFL</td>\n",
       "      <td>28542</td>\n",
       "      <td>2</td>\n",
       "      <td>28608</td>\n",
       "      <td>Branchiostoma floridae (Florida lancelet) (Amp...</td>\n",
       "      <td>684.0</td>\n",
       "      <td>Branchiostoma floridae</td>\n",
       "      <td>Florida lancelet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UP000001940</td>\n",
       "      <td>6239</td>\n",
       "      <td>CAEEL</td>\n",
       "      <td>19986</td>\n",
       "      <td>8309</td>\n",
       "      <td>28507</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>797.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteome_id  tax_id oscode  n_main_fasta_canonical  \\\n",
       "0  UP000007062    7165  ANOGA                   12553   \n",
       "5  UP000007241  684364  BATDJ                    8610   \n",
       "6  UP000009136    9913  BOVIN                   23774   \n",
       "8  UP000001554    7739  BRAFL                   28542   \n",
       "9  UP000001940    6239  CAEEL                   19986   \n",
       "\n",
       "   n_additional_fasta_isoforms  n_gene2acc  \\\n",
       "0                          971       13619   \n",
       "5                            0        8685   \n",
       "6                        14534       38438   \n",
       "8                            2       28608   \n",
       "9                         8309       28507   \n",
       "\n",
       "                                        species_name  \\\n",
       "0       Anopheles gambiae (African malaria mosquito)   \n",
       "5  Batrachochytrium dendrobatidis (strain JAM81 /...   \n",
       "6                                Bos taurus (Bovine)   \n",
       "8  Branchiostoma floridae (Florida lancelet) (Amp...   \n",
       "9                             Caenorhabditis elegans   \n",
       "\n",
       "   divergence_from_human_mya                 scientific_name  \\\n",
       "0                      797.0               Anopheles gambiae   \n",
       "5                     1105.0  Batrachochytrium dendrobatidis   \n",
       "6                       96.0                      Bos taurus   \n",
       "8                      684.0          Branchiostoma floridae   \n",
       "9                      797.0          Caenorhabditis elegans   \n",
       "\n",
       "       common_name_or_strain  \n",
       "0   African malaria mosquito  \n",
       "5  strain JAM81 / FGSC 10211  \n",
       "6                     Bovine  \n",
       "8           Florida lancelet  \n",
       "9                        NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opisthokonts_not_human = opisthokonts.query('scientific_name != \"Homo sapiens\"')\n",
    "print(opisthokonts_not_human.shape)\n",
    "opisthokonts_not_human.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use ENSEMBL Rest API to get homologous TFs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Actually run it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source__uniprot_id</th>\n",
       "      <th>source__id_type</th>\n",
       "      <th>source__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P29973</td>\n",
       "      <td>UniProtKB-ID</td>\n",
       "      <td>CNGA1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Name</td>\n",
       "      <td>CNGA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>CNCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>CNCG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P29973</td>\n",
       "      <td>GI</td>\n",
       "      <td>2506302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source__uniprot_id source__id_type source__db_id\n",
       "0             P29973    UniProtKB-ID   CNGA1_HUMAN\n",
       "1             P29973       Gene_Name         CNGA1\n",
       "2             P29973    Gene_Synonym          CNCG\n",
       "3             P29973    Gene_Synonym         CNCG1\n",
       "4             P29973              GI       2506302"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source__uniprot_id</th>\n",
       "      <th>source__id_type</th>\n",
       "      <th>source__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2055</th>\n",
       "      <td>P08100</td>\n",
       "      <td>GI</td>\n",
       "      <td>1236137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>A0A1B0GVU3</td>\n",
       "      <td>CRC64</td>\n",
       "      <td>89611212D656D06D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1609</th>\n",
       "      <td>P62873</td>\n",
       "      <td>neXtProt</td>\n",
       "      <td>NX_P62873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>963</th>\n",
       "      <td>Q9UMX6</td>\n",
       "      <td>GeneReviews</td>\n",
       "      <td>GUCA1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2765</th>\n",
       "      <td>P51160</td>\n",
       "      <td>BioMuta</td>\n",
       "      <td>PDE6C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2339</th>\n",
       "      <td>C9J7V6</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000133256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2998</th>\n",
       "      <td>A6PVH5</td>\n",
       "      <td>CRC64</td>\n",
       "      <td>176DA167CA68CA96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2564</th>\n",
       "      <td>P43080</td>\n",
       "      <td>GI</td>\n",
       "      <td>623405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078</th>\n",
       "      <td>Q9UBF6</td>\n",
       "      <td>EMBL-CDS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1120</th>\n",
       "      <td>Q9UBF6</td>\n",
       "      <td>CRC64</td>\n",
       "      <td>CE1E6CAC940C8257</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source__uniprot_id source__id_type     source__db_id\n",
       "2055             P08100              GI           1236137\n",
       "564          A0A1B0GVU3           CRC64  89611212D656D06D\n",
       "1609             P62873        neXtProt         NX_P62873\n",
       "963              Q9UMX6     GeneReviews            GUCA1B\n",
       "2765             P51160         BioMuta             PDE6C\n",
       "2339             C9J7V6         Ensembl   ENSG00000133256\n",
       "2998             A6PVH5           CRC64  176DA167CA68CA96\n",
       "2564             P43080              GI            623405\n",
       "1078             Q9UBF6        EMBL-CDS                 -\n",
       "1120             Q9UBF6           CRC64  CE1E6CAC940C8257"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Separate out samples with ensg ids and not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2981, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source__uniprot_id</th>\n",
       "      <th>source__id_type</th>\n",
       "      <th>source__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P29973</td>\n",
       "      <td>UniProtKB-ID</td>\n",
       "      <td>CNGA1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Name</td>\n",
       "      <td>CNGA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>CNCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Gene_Synonym</td>\n",
       "      <td>CNCG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P29973</td>\n",
       "      <td>GI</td>\n",
       "      <td>2506302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source__uniprot_id source__id_type source__db_id\n",
       "0             P29973    UniProtKB-ID   CNGA1_HUMAN\n",
       "1             P29973       Gene_Name         CNGA1\n",
       "2             P29973    Gene_Synonym          CNCG\n",
       "3             P29973    Gene_Synonym         CNCG1\n",
       "4             P29973              GI       2506302"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg_rows = visual.source__db_id.str.startswith(\"ENSG0\")\n",
    "visual_non_ensg = visual.loc[~ensg_rows]\n",
    "print(visual_non_ensg.shape)\n",
    "visual_non_ensg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source__uniprot_id</th>\n",
       "      <th>source__id_type</th>\n",
       "      <th>source__db_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>P29973</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>P16520</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>P11488</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000114349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Q02846</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000132518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>P35913</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000133256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    source__uniprot_id source__id_type    source__db_id\n",
       "42              P29973         Ensembl  ENSG00000198515\n",
       "119             P16520         Ensembl  ENSG00000111664\n",
       "212             P11488         Ensembl  ENSG00000114349\n",
       "283             Q02846         Ensembl  ENSG00000132518\n",
       "370             P35913         Ensembl  ENSG00000133256"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual_ensg = visual.loc[ensg_rows]\n",
    "print(visual_ensg.shape)\n",
    "visual_ensg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.error import HTTPError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|          | 0/72 [00:00<?, ?it/s]\u001b[A\n",
      "  1%|▏         | 1/72 [00:11<13:38, 11.53s/it]\u001b[A\n",
      "  3%|▎         | 2/72 [00:35<17:51, 15.31s/it]\u001b[A\n",
      "  4%|▍         | 3/72 [00:51<17:40, 15.37s/it]\u001b[A\n",
      "  6%|▌         | 4/72 [01:05<17:06, 15.10s/it]\u001b[A\n",
      "  7%|▋         | 5/72 [01:23<17:38, 15.80s/it]\u001b[A\n",
      "  8%|▊         | 6/72 [01:37<16:45, 15.24s/it]\u001b[A\n",
      " 10%|▉         | 7/72 [01:53<16:52, 15.58s/it]\u001b[A\n",
      " 11%|█         | 8/72 [01:58<13:13, 12.40s/it]\u001b[A\n",
      " 12%|█▎        | 9/72 [02:04<10:54, 10.38s/it]\u001b[A\n",
      " 14%|█▍        | 10/72 [02:20<12:43, 12.31s/it]\u001b[A\n",
      " 15%|█▌        | 11/72 [02:43<15:33, 15.30s/it]\u001b[A\n",
      " 17%|█▋        | 12/72 [02:59<15:44, 15.73s/it]\u001b[A\n",
      " 18%|█▊        | 13/72 [03:04<12:13, 12.44s/it]\u001b[A\n",
      " 19%|█▉        | 14/72 [03:10<09:59, 10.33s/it]\u001b[A\n",
      " 21%|██        | 15/72 [03:29<12:24, 13.06s/it]\u001b[A\n",
      " 22%|██▏       | 16/72 [03:30<08:54,  9.54s/it]\u001b[A\n",
      " 24%|██▎       | 17/72 [03:39<08:27,  9.24s/it]\u001b[A\n",
      " 25%|██▌       | 18/72 [03:54<09:56, 11.05s/it]\u001b[A\n",
      " 26%|██▋       | 19/72 [04:10<11:00, 12.47s/it]\u001b[A\n",
      " 28%|██▊       | 20/72 [04:19<09:50, 11.36s/it]\u001b[A\n",
      " 29%|██▉       | 21/72 [04:25<08:19,  9.80s/it]\u001b[A\n",
      " 31%|███       | 22/72 [04:41<09:47, 11.75s/it]\u001b[A\n",
      " 32%|███▏      | 23/72 [05:00<11:25, 13.99s/it]\u001b[A\n",
      " 33%|███▎      | 24/72 [05:11<10:20, 12.94s/it]\u001b[A\n",
      " 35%|███▍      | 25/72 [05:15<08:07, 10.38s/it]\u001b[A\n",
      " 36%|███▌      | 26/72 [05:20<06:41,  8.72s/it]\u001b[A\n",
      " 38%|███▊      | 27/72 [05:26<05:50,  7.80s/it]\u001b[A\n",
      " 39%|███▉      | 28/72 [05:29<04:45,  6.48s/it]\u001b[A\n",
      " 40%|████      | 29/72 [05:32<03:50,  5.37s/it]\u001b[A\n",
      " 42%|████▏     | 30/72 [05:37<03:35,  5.14s/it]\u001b[A\n",
      " 43%|████▎     | 31/72 [05:50<05:10,  7.56s/it]\u001b[A\n",
      " 44%|████▍     | 32/72 [05:53<04:10,  6.27s/it]\u001b[A\n",
      " 46%|████▌     | 33/72 [06:16<07:15, 11.16s/it]\u001b[A\n",
      " 47%|████▋     | 34/72 [06:19<05:30,  8.71s/it]\u001b[A\n",
      " 49%|████▊     | 35/72 [06:22<04:28,  7.26s/it]\u001b[A\n",
      " 50%|█████     | 36/72 [06:25<03:34,  5.95s/it]\u001b[A\n",
      " 51%|█████▏    | 37/72 [06:28<02:52,  4.94s/it]\u001b[A\n",
      " 53%|█████▎    | 38/72 [06:48<05:26,  9.61s/it]\u001b[A\n",
      " 54%|█████▍    | 39/72 [07:01<05:44, 10.44s/it]\u001b[A\n",
      " 56%|█████▌    | 40/72 [07:15<06:08, 11.52s/it]\u001b[A\n",
      " 57%|█████▋    | 41/72 [07:31<06:40, 12.91s/it]\u001b[A\n",
      " 58%|█████▊    | 42/72 [07:35<05:06, 10.21s/it]\u001b[A\n",
      " 60%|█████▉    | 43/72 [07:48<05:22, 11.13s/it]\u001b[A\n",
      " 61%|██████    | 44/72 [07:53<04:20,  9.32s/it]\u001b[A\n",
      " 62%|██████▎   | 45/72 [07:54<03:06,  6.89s/it]\u001b[A\n",
      " 64%|██████▍   | 46/72 [08:00<02:45,  6.35s/it]\u001b[A\n",
      " 65%|██████▌   | 47/72 [08:04<02:24,  5.77s/it]\u001b[A\n",
      " 67%|██████▋   | 48/72 [08:07<02:02,  5.08s/it]\u001b[A\n",
      " 68%|██████▊   | 49/72 [08:10<01:39,  4.34s/it]\u001b[A\n",
      " 69%|██████▉   | 50/72 [08:14<01:34,  4.31s/it]\u001b[A\n",
      " 71%|███████   | 51/72 [08:18<01:23,  3.98s/it]\u001b[A\n",
      " 72%|███████▏  | 52/72 [08:21<01:16,  3.83s/it]\u001b[A\n",
      " 74%|███████▎  | 53/72 [08:24<01:09,  3.64s/it]\u001b[A\n",
      " 75%|███████▌  | 54/72 [08:30<01:17,  4.33s/it]\u001b[A\n",
      " 76%|███████▋  | 55/72 [08:33<01:04,  3.82s/it]\u001b[A\n",
      " 78%|███████▊  | 56/72 [08:39<01:12,  4.55s/it]\u001b[A\n",
      " 79%|███████▉  | 57/72 [08:40<00:53,  3.55s/it]\u001b[A\n",
      " 81%|████████  | 58/72 [08:44<00:48,  3.48s/it]\u001b[A\n",
      " 82%|████████▏ | 59/72 [08:57<01:22,  6.38s/it]\u001b[A\n",
      " 83%|████████▎ | 60/72 [09:10<01:42,  8.57s/it]\u001b[A\n",
      " 85%|████████▍ | 61/72 [09:15<01:21,  7.43s/it]\u001b[A\n",
      " 86%|████████▌ | 62/72 [09:19<01:03,  6.38s/it]\u001b[A\n",
      " 88%|████████▊ | 63/72 [09:20<00:43,  4.83s/it]\u001b[A\n",
      " 89%|████████▉ | 64/72 [09:23<00:33,  4.23s/it]\u001b[A\n",
      " 90%|█████████ | 65/72 [09:27<00:28,  4.06s/it]\u001b[A\n",
      " 92%|█████████▏| 66/72 [09:28<00:19,  3.21s/it]\u001b[A\n",
      " 93%|█████████▎| 67/72 [09:33<00:18,  3.70s/it]\u001b[A\n",
      " 94%|█████████▍| 68/72 [09:36<00:13,  3.42s/it]\u001b[A\n",
      " 96%|█████████▌| 69/72 [09:38<00:09,  3.14s/it]\u001b[A\n",
      " 97%|█████████▋| 70/72 [09:42<00:06,  3.42s/it]\u001b[A\n",
      " 99%|█████████▊| 71/72 [09:45<00:03,  3.21s/it]\u001b[A\n",
      "100%|██████████| 72/72 [09:49<00:00,  8.19s/it]\u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17828, 16)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "<ipython-input-52-1b665e2cc1da>:33: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  visual_opsithokonts_ensembl = pd.concat(dfs, ignore_index=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>target__perc_pos</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>50.5929</td>\n",
       "      <td>62.8458</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000183862</td>\n",
       "      <td>57.8313</td>\n",
       "      <td>71.8373</td>\n",
       "      <td>ENSP00000328478</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>18.1818</td>\n",
       "      <td>35.9684</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000055118</td>\n",
       "      <td>11.9068</td>\n",
       "      <td>23.5548</td>\n",
       "      <td>ENSP00000262186</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>17.3913</td>\n",
       "      <td>35.1779</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000143630</td>\n",
       "      <td>17.0543</td>\n",
       "      <td>34.4961</td>\n",
       "      <td>ENSP00000357342</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>55.0725</td>\n",
       "      <td>67.0619</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000144191</td>\n",
       "      <td>60.2305</td>\n",
       "      <td>73.3429</td>\n",
       "      <td>ENSP00000272602</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>24.7694</td>\n",
       "      <td>40.8432</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000170289</td>\n",
       "      <td>23.2386</td>\n",
       "      <td>38.3189</td>\n",
       "      <td>ENSP00000316605</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  dn_ds    method_link_type       source__id  source__perc_id  \\\n",
       "0   NaN  ENSEMBL_PARALOGUES  ENSG00000198515          50.5929   \n",
       "1   NaN  ENSEMBL_PARALOGUES  ENSG00000198515          18.1818   \n",
       "2   NaN  ENSEMBL_PARALOGUES  ENSG00000198515          17.3913   \n",
       "3   NaN  ENSEMBL_PARALOGUES  ENSG00000198515          55.0725   \n",
       "4   NaN  ENSEMBL_PARALOGUES  ENSG00000198515          24.7694   \n",
       "\n",
       "   source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "0           62.8458    ENSP00000384264    homo_sapiens              9606   \n",
       "1           35.9684    ENSP00000384264    homo_sapiens              9606   \n",
       "2           35.1779    ENSP00000384264    homo_sapiens              9606   \n",
       "3           67.0619    ENSP00000384264    homo_sapiens              9606   \n",
       "4           40.8432    ENSP00000384264    homo_sapiens              9606   \n",
       "\n",
       "        target__id  target__perc_id  target__perc_pos target__protein_id  \\\n",
       "0  ENSG00000183862          57.8313           71.8373    ENSP00000328478   \n",
       "1  ENSG00000055118          11.9068           23.5548    ENSP00000262186   \n",
       "2  ENSG00000143630          17.0543           34.4961    ENSP00000357342   \n",
       "3  ENSG00000144191          60.2305           73.3429    ENSP00000272602   \n",
       "4  ENSG00000170289          23.2386           38.3189    ENSP00000316605   \n",
       "\n",
       "  target__species  target__taxon_id taxonomy_level           type  \n",
       "0    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "1    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "2    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "3    homo_sapiens              9606      Bilateria  other_paralog  \n",
       "4    homo_sapiens              9606      Bilateria  other_paralog  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "dfs = []\n",
    "\n",
    "\n",
    "\n",
    "def single_homology_to_series(homology, ignore_fields=['align_seq', 'cigar_line', 'target', 'source']):\n",
    "\n",
    "    homology_for_series = {}\n",
    "    for key, value in homology.items():\n",
    "        if not key in ignore_fields:\n",
    "            homology_for_series[key] = value\n",
    "        if key in ('target', 'source'):\n",
    "            for k, v in value.items():\n",
    "                if not k in ignore_fields:\n",
    "                    homology_for_series[f\"{key}__{k}\"] = v\n",
    "\n",
    "    series = pd.Series(homology_for_series)\n",
    "    return series\n",
    "\n",
    "\n",
    "for ensembl_id in tqdm(visual_ensg.source__db_id):\n",
    "    sleep(1)\n",
    "    try:\n",
    "        response = ensembl_rest.homology_ensemblgene(ensembl_id, target_taxon=opisthokonts_not_human.tax_id, cigar_line=False, sequence=None)\n",
    "    except HTTPError:\n",
    "        # Probably a 503 error, meaning server is busy, so wait 2 seconds and try again\n",
    "        sleep(2)\n",
    "        response = ensembl_rest.homology_ensemblgene(ensembl_id, target_taxon=opisthokonts_not_human.tax_id, cigar_line=False, sequence=None)\n",
    "\n",
    "    data = response['data']\n",
    "    homologies = data[0]['homologies']\n",
    "    df = pd.DataFrame(map(single_homology_to_series, homologies))\n",
    "    dfs.append(df)\n",
    "visual_opsithokonts_ensembl = pd.concat(dfs, ignore_index=True)\n",
    "print(visual_opsithokonts_ensembl.shape)\n",
    "visual_opsithokonts_ensembl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>target__perc_pos</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>50.5929</td>\n",
       "      <td>62.8458</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000183862</td>\n",
       "      <td>57.8313</td>\n",
       "      <td>71.8373</td>\n",
       "      <td>ENSP00000328478</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>18.1818</td>\n",
       "      <td>35.9684</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000055118</td>\n",
       "      <td>11.9068</td>\n",
       "      <td>23.5548</td>\n",
       "      <td>ENSP00000262186</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>17.3913</td>\n",
       "      <td>35.1779</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000143630</td>\n",
       "      <td>17.0543</td>\n",
       "      <td>34.4961</td>\n",
       "      <td>ENSP00000357342</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>55.0725</td>\n",
       "      <td>67.0619</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000144191</td>\n",
       "      <td>60.2305</td>\n",
       "      <td>73.3429</td>\n",
       "      <td>ENSP00000272602</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_PARALOGUES</td>\n",
       "      <td>ENSG00000198515</td>\n",
       "      <td>24.7694</td>\n",
       "      <td>40.8432</td>\n",
       "      <td>ENSP00000384264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000170289</td>\n",
       "      <td>23.2386</td>\n",
       "      <td>38.3189</td>\n",
       "      <td>ENSP00000316605</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>other_paralog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17823</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000114349</td>\n",
       "      <td>88.8571</td>\n",
       "      <td>95.7143</td>\n",
       "      <td>ENSP00000232461</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSSTUG00000023497</td>\n",
       "      <td>88.8571</td>\n",
       "      <td>95.7143</td>\n",
       "      <td>ENSSTUP00000055369</td>\n",
       "      <td>salmo_trutta</td>\n",
       "      <td>8032</td>\n",
       "      <td>Gnathostomata</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17824</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000114349</td>\n",
       "      <td>91.4286</td>\n",
       "      <td>96.5714</td>\n",
       "      <td>ENSP00000232461</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSMALG00000001144</td>\n",
       "      <td>91.4286</td>\n",
       "      <td>96.5714</td>\n",
       "      <td>ENSMALP00000001558</td>\n",
       "      <td>monopterus_albus</td>\n",
       "      <td>43700</td>\n",
       "      <td>Gnathostomata</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17825</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000114349</td>\n",
       "      <td>76.0000</td>\n",
       "      <td>80.5714</td>\n",
       "      <td>ENSP00000232461</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSCSEG00000000424</td>\n",
       "      <td>90.7850</td>\n",
       "      <td>96.2457</td>\n",
       "      <td>ENSCSEP00000000602</td>\n",
       "      <td>cynoglossus_semilaevis</td>\n",
       "      <td>244447</td>\n",
       "      <td>Gnathostomata</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17826</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000114349</td>\n",
       "      <td>30.8571</td>\n",
       "      <td>33.4286</td>\n",
       "      <td>ENSP00000232461</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSPMAG00000004273</td>\n",
       "      <td>87.8049</td>\n",
       "      <td>95.1219</td>\n",
       "      <td>ENSPMAP00000004667</td>\n",
       "      <td>petromyzon_marinus</td>\n",
       "      <td>7757</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17827</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000114349</td>\n",
       "      <td>72.0000</td>\n",
       "      <td>80.5714</td>\n",
       "      <td>ENSP00000232461</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSEBUG00000009856</td>\n",
       "      <td>83.4437</td>\n",
       "      <td>93.3775</td>\n",
       "      <td>ENSEBUP00000015660</td>\n",
       "      <td>eptatretus_burgeri</td>\n",
       "      <td>7764</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17828 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dn_ds     method_link_type       source__id  source__perc_id  \\\n",
       "0       NaN   ENSEMBL_PARALOGUES  ENSG00000198515          50.5929   \n",
       "1       NaN   ENSEMBL_PARALOGUES  ENSG00000198515          18.1818   \n",
       "2       NaN   ENSEMBL_PARALOGUES  ENSG00000198515          17.3913   \n",
       "3       NaN   ENSEMBL_PARALOGUES  ENSG00000198515          55.0725   \n",
       "4       NaN   ENSEMBL_PARALOGUES  ENSG00000198515          24.7694   \n",
       "...     ...                  ...              ...              ...   \n",
       "17823   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000114349          88.8571   \n",
       "17824   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000114349          91.4286   \n",
       "17825   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000114349          76.0000   \n",
       "17826   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000114349          30.8571   \n",
       "17827   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000114349          72.0000   \n",
       "\n",
       "       source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "0               62.8458    ENSP00000384264    homo_sapiens              9606   \n",
       "1               35.9684    ENSP00000384264    homo_sapiens              9606   \n",
       "2               35.1779    ENSP00000384264    homo_sapiens              9606   \n",
       "3               67.0619    ENSP00000384264    homo_sapiens              9606   \n",
       "4               40.8432    ENSP00000384264    homo_sapiens              9606   \n",
       "...                 ...                ...             ...               ...   \n",
       "17823           95.7143    ENSP00000232461    homo_sapiens              9606   \n",
       "17824           96.5714    ENSP00000232461    homo_sapiens              9606   \n",
       "17825           80.5714    ENSP00000232461    homo_sapiens              9606   \n",
       "17826           33.4286    ENSP00000232461    homo_sapiens              9606   \n",
       "17827           80.5714    ENSP00000232461    homo_sapiens              9606   \n",
       "\n",
       "               target__id  target__perc_id  target__perc_pos  \\\n",
       "0         ENSG00000183862          57.8313           71.8373   \n",
       "1         ENSG00000055118          11.9068           23.5548   \n",
       "2         ENSG00000143630          17.0543           34.4961   \n",
       "3         ENSG00000144191          60.2305           73.3429   \n",
       "4         ENSG00000170289          23.2386           38.3189   \n",
       "...                   ...              ...               ...   \n",
       "17823  ENSSTUG00000023497          88.8571           95.7143   \n",
       "17824  ENSMALG00000001144          91.4286           96.5714   \n",
       "17825  ENSCSEG00000000424          90.7850           96.2457   \n",
       "17826  ENSPMAG00000004273          87.8049           95.1219   \n",
       "17827  ENSEBUG00000009856          83.4437           93.3775   \n",
       "\n",
       "       target__protein_id         target__species  target__taxon_id  \\\n",
       "0         ENSP00000328478            homo_sapiens              9606   \n",
       "1         ENSP00000262186            homo_sapiens              9606   \n",
       "2         ENSP00000357342            homo_sapiens              9606   \n",
       "3         ENSP00000272602            homo_sapiens              9606   \n",
       "4         ENSP00000316605            homo_sapiens              9606   \n",
       "...                   ...                     ...               ...   \n",
       "17823  ENSSTUP00000055369            salmo_trutta              8032   \n",
       "17824  ENSMALP00000001558        monopterus_albus             43700   \n",
       "17825  ENSCSEP00000000602  cynoglossus_semilaevis            244447   \n",
       "17826  ENSPMAP00000004667      petromyzon_marinus              7757   \n",
       "17827  ENSEBUP00000015660      eptatretus_burgeri              7764   \n",
       "\n",
       "      taxonomy_level               type  \n",
       "0          Bilateria      other_paralog  \n",
       "1          Bilateria      other_paralog  \n",
       "2          Bilateria      other_paralog  \n",
       "3          Bilateria      other_paralog  \n",
       "4          Bilateria      other_paralog  \n",
       "...              ...                ...  \n",
       "17823  Gnathostomata  ortholog_one2many  \n",
       "17824  Gnathostomata   ortholog_one2one  \n",
       "17825  Gnathostomata   ortholog_one2one  \n",
       "17826     Vertebrata   ortholog_one2one  \n",
       "17827     Vertebrata   ortholog_one2one  \n",
       "\n",
       "[17828 rows x 16 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual_opsithokonts_ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "homo_sapiens                    815\n",
       "sinocyclocheilus_rhinocerous    183\n",
       "cyprinus_carpio_germanmirror    178\n",
       "cyprinus_carpio_huanghe         176\n",
       "cyprinus_carpio_hebaored        174\n",
       "                               ... \n",
       "sorex_araneus                    37\n",
       "ciona_savignyi                   35\n",
       "chrysolophus_pictus              35\n",
       "choloepus_hoffmanni              23\n",
       "saccharomyces_cerevisiae          6\n",
       "Name: target__species, Length: 240, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual_opsithokonts_ensembl.target__species.value_counts() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dn_ds</th>\n",
       "      <th>method_link_type</th>\n",
       "      <th>source__id</th>\n",
       "      <th>source__perc_id</th>\n",
       "      <th>source__perc_pos</th>\n",
       "      <th>source__protein_id</th>\n",
       "      <th>source__species</th>\n",
       "      <th>source__taxon_id</th>\n",
       "      <th>target__id</th>\n",
       "      <th>target__perc_id</th>\n",
       "      <th>target__perc_pos</th>\n",
       "      <th>target__protein_id</th>\n",
       "      <th>target__species</th>\n",
       "      <th>target__taxon_id</th>\n",
       "      <th>taxonomy_level</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "      <td>40.8824</td>\n",
       "      <td>61.1765</td>\n",
       "      <td>ENSP00000229264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>32.8605</td>\n",
       "      <td>49.1726</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2197</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "      <td>40.8824</td>\n",
       "      <td>61.1765</td>\n",
       "      <td>ENSP00000229264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>32.8605</td>\n",
       "      <td>49.1726</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7038</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "      <td>40.8824</td>\n",
       "      <td>61.1765</td>\n",
       "      <td>ENSP00000229264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>32.8605</td>\n",
       "      <td>49.1726</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12119</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "      <td>40.8824</td>\n",
       "      <td>61.1765</td>\n",
       "      <td>ENSP00000229264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>32.8605</td>\n",
       "      <td>49.1726</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12542</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "      <td>40.8824</td>\n",
       "      <td>61.1765</td>\n",
       "      <td>ENSP00000229264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>32.8605</td>\n",
       "      <td>49.1726</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17483</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSEMBL_ORTHOLOGUES</td>\n",
       "      <td>ENSG00000111664</td>\n",
       "      <td>40.8824</td>\n",
       "      <td>61.1765</td>\n",
       "      <td>ENSP00000229264</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>32.8605</td>\n",
       "      <td>49.1726</td>\n",
       "      <td>YOR212W</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      dn_ds     method_link_type       source__id  source__perc_id  \\\n",
       "667     NaN  ENSEMBL_ORTHOLOGUES  ENSG00000111664          40.8824   \n",
       "2197    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000111664          40.8824   \n",
       "7038    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000111664          40.8824   \n",
       "12119   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000111664          40.8824   \n",
       "12542   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000111664          40.8824   \n",
       "17483   NaN  ENSEMBL_ORTHOLOGUES  ENSG00000111664          40.8824   \n",
       "\n",
       "       source__perc_pos source__protein_id source__species  source__taxon_id  \\\n",
       "667             61.1765    ENSP00000229264    homo_sapiens              9606   \n",
       "2197            61.1765    ENSP00000229264    homo_sapiens              9606   \n",
       "7038            61.1765    ENSP00000229264    homo_sapiens              9606   \n",
       "12119           61.1765    ENSP00000229264    homo_sapiens              9606   \n",
       "12542           61.1765    ENSP00000229264    homo_sapiens              9606   \n",
       "17483           61.1765    ENSP00000229264    homo_sapiens              9606   \n",
       "\n",
       "      target__id  target__perc_id  target__perc_pos target__protein_id  \\\n",
       "667      YOR212W          32.8605           49.1726            YOR212W   \n",
       "2197     YOR212W          32.8605           49.1726            YOR212W   \n",
       "7038     YOR212W          32.8605           49.1726            YOR212W   \n",
       "12119    YOR212W          32.8605           49.1726            YOR212W   \n",
       "12542    YOR212W          32.8605           49.1726            YOR212W   \n",
       "17483    YOR212W          32.8605           49.1726            YOR212W   \n",
       "\n",
       "                target__species  target__taxon_id taxonomy_level  \\\n",
       "667    saccharomyces_cerevisiae              4932   Opisthokonta   \n",
       "2197   saccharomyces_cerevisiae              4932   Opisthokonta   \n",
       "7038   saccharomyces_cerevisiae              4932   Opisthokonta   \n",
       "12119  saccharomyces_cerevisiae              4932   Opisthokonta   \n",
       "12542  saccharomyces_cerevisiae              4932   Opisthokonta   \n",
       "17483  saccharomyces_cerevisiae              4932   Opisthokonta   \n",
       "\n",
       "                    type  \n",
       "667    ortholog_one2many  \n",
       "2197   ortholog_one2many  \n",
       "7038   ortholog_one2many  \n",
       "12119  ortholog_one2many  \n",
       "12542  ortholog_one2many  \n",
       "17483  ortholog_one2many  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual_opsithokonts_ensembl.query('target__species == \"saccharomyces_cerevisiae\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The only gene here is [STE4 in Yeast](https://www.yeastgenome.org/locus/S000005738) which corresponds to [GNB3 in human](https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000111664;r=12:6839954-6847393)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write to disk!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tfs_opisthokonts = pd.concat([tfs_opsithokonts_symbol, tfs_opsithokonts_ensembl])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "visual_opsithokonts_ensembl.to_csv('opisthokont_not_human_visual_system_ensembl_compara.csv', index=False)\n",
    "visual_opsithokonts_ensembl.to_parquet('opisthokont_not_human_visual_system_ensembl_compara.parquet', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls -lha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:khtools--encodings--compare-cli]",
   "language": "python",
   "name": "conda-env-khtools--encodings--compare-cli-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
